{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Tree Data Preparation"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "This notebook demonstrates how to prepare the input for FiTree using the AML trees from [Morita et al., 2020](https://www.nature.com/articles/s41467-020-19119-8). "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### 1. Load necessary packages"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "from pyfaidx import Fasta\n",
    "import matplotlib.pyplot as plt\n",
    "from fitree import (\n",
    "    Subclone, \n",
    "    TumorTree, \n",
    "    TumorTreeCohort, \n",
    "    save_cohort_to_json, \n",
    "    load_cohort_from_json,\n",
    "    plot_tree,\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### 2. Load the dataframe of tumor mutation trees"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The following columns are necessary for the tree preparation:\n",
    "- Patient: patient labels in text\n",
    "- Mutation: mutation labels in text\n",
    "- Patient_ID: patient IDs\n",
    "- Tree_ID: tree IDs (one patient can have multiple trees)\n",
    "- Node_ID: IDs of each node/subclone in the tree\n",
    "- Mutation_ID: mutation IDs\n",
    "- Parent_ID: IDs of the parent nodes (Parent_ID = 0 means that the parent is the root)\n",
    "- Age: ages of the patients\n",
    "- Total_Cells: total number of tumor cells. This value should be estimated from clinical data. For AML, the formula is\n",
    "\n",
    "$$\n",
    "\\begin{split}\n",
    "    & \\quad \\text{PB\\_WBC} \\times \\text{PB\\_Blast} \\times \\text{Average adult blood volume} \\\\\n",
    "    & + \\text{BM\\_Blast} \\times \\text{Average adult bone marrow mononuclear cell count}\n",
    "\\end{split}\n",
    "$$\n",
    "\n",
    "\n",
    "- Sequenced_Cells: number of cells attached to the node/subclone"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "AML_tree_df_combined = pd.read_csv('../data/AML_tree_df_Morita_2020.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Patient</th>\n",
       "      <th>Mutation</th>\n",
       "      <th>Node_ID</th>\n",
       "      <th>Mutation_ID</th>\n",
       "      <th>Parent_ID</th>\n",
       "      <th>Patient_ID</th>\n",
       "      <th>Tree_ID</th>\n",
       "      <th>Age</th>\n",
       "      <th>Total_Cells</th>\n",
       "      <th>Sequenced_Cells</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AML-01-001</td>\n",
       "      <td>FLT3</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>66</td>\n",
       "      <td>9.490000e+11</td>\n",
       "      <td>390</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>AML-01-001</td>\n",
       "      <td>WT1</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>66</td>\n",
       "      <td>9.490000e+11</td>\n",
       "      <td>3378</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>AML-01-001</td>\n",
       "      <td>NPM1</td>\n",
       "      <td>3</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>66</td>\n",
       "      <td>9.490000e+11</td>\n",
       "      <td>7306</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>AML-02-001</td>\n",
       "      <td>DNMT3A</td>\n",
       "      <td>1</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>84</td>\n",
       "      <td>2.090000e+11</td>\n",
       "      <td>529</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>AML-02-001</td>\n",
       "      <td>SF3B1</td>\n",
       "      <td>2</td>\n",
       "      <td>16</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>84</td>\n",
       "      <td>2.090000e+11</td>\n",
       "      <td>262</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      Patient Mutation  Node_ID  Mutation_ID  Parent_ID  Patient_ID  Tree_ID  \\\n",
       "0  AML-01-001     FLT3        1            0          0           0        0   \n",
       "1  AML-01-001      WT1        2            2          1           0        0   \n",
       "2  AML-01-001     NPM1        3            1          2           0        0   \n",
       "3  AML-02-001   DNMT3A        1            3          0           1        0   \n",
       "4  AML-02-001    SF3B1        2           16          1           1        0   \n",
       "\n",
       "   Age   Total_Cells  Sequenced_Cells  \n",
       "0   66  9.490000e+11              390  \n",
       "1   66  9.490000e+11             3378  \n",
       "2   66  9.490000e+11             7306  \n",
       "3   84  2.090000e+11              529  \n",
       "4   84  2.090000e+11              262  "
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "AML_tree_df_combined.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### 3. Process data and estimate static parameters"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We assume that evolution begins once the HSC reserve fills, average around 18 years, which is subtracted from the patientsâ€™ ages at diagnosis to get the sampling times ts of the tumors."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "AML_tree_df_combined[\"Sampling_Time\"] = AML_tree_df_combined[\"Age\"] - 18"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The population of human hematopoietic stem cells (HSCs) grows from birth through adolescence and then stabilizes, resulting in a relatively static population with an average of approximately $C_0 = 10^5$ wild-type cells, which we model as the root in our framework."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "C_0 = 1e5"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The maximum observation time $t_{\\max}$ is then set as the maximum of the sampling times."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "t_max = np.max(AML_tree_df_combined[\"Sampling_Time\"])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We take the common death rate $\\beta = 1.0$ to match roughly the average birth rate of HSCs."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "common_beta = 1.0"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The lifetime risk of AML is approximately 0.5% according to [SEER](https://seer.cancer.gov/statfacts/html/amyl.html)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "lifetime_risk = 0.005"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The scaling factor $C_{\\text{sampling}}$ is taken to be the mean of the tumor sizes."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "C_sampling = np.mean(AML_tree_df_combined[\"Total_Cells\"].unique()) "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "455248214285.7143"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "C_sampling"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### 3.1 Determine mutation rates"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We first load the variants information provided by [Morita et al., 2020](https://www.nature.com/articles/s41467-020-19119-8). This dataframe lists all the variants detected in the cohort and indicates whether they are SNVs."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "AML_variants = pd.read_csv('../data/AML_variants_Morita_2020.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample ID</th>\n",
       "      <th>gene</th>\n",
       "      <th>chr</th>\n",
       "      <th>start</th>\n",
       "      <th>ref allele</th>\n",
       "      <th>alt allele</th>\n",
       "      <th>amino acid change</th>\n",
       "      <th>function</th>\n",
       "      <th>exonic function</th>\n",
       "      <th>mutated %</th>\n",
       "      <th>scDNA-seq VAF</th>\n",
       "      <th>WT\\n(No. of cells)</th>\n",
       "      <th>Het\\n(No. of cells)</th>\n",
       "      <th>Homo\\n(No. of cells)</th>\n",
       "      <th>Missing\\n(No. of cells)</th>\n",
       "      <th>validated</th>\n",
       "      <th>validation method</th>\n",
       "      <th>bulk VAF</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AML-01-001</td>\n",
       "      <td>FLT3</td>\n",
       "      <td>13</td>\n",
       "      <td>28608259</td>\n",
       "      <td>X</td>\n",
       "      <td>XAACTCTAAATTTTCTCTTGGAAACTCCCATTTGAGATCATATTCA...</td>\n",
       "      <td>p.Y599_D600delinsX</td>\n",
       "      <td>exonic</td>\n",
       "      <td>stopgain</td>\n",
       "      <td>0.9352</td>\n",
       "      <td>0.7949</td>\n",
       "      <td>239</td>\n",
       "      <td>8444</td>\n",
       "      <td>2190</td>\n",
       "      <td>498</td>\n",
       "      <td>yes</td>\n",
       "      <td>bulk-NGS</td>\n",
       "      <td>0.1186</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>AML-01-001</td>\n",
       "      <td>NPM1</td>\n",
       "      <td>5</td>\n",
       "      <td>170837544</td>\n",
       "      <td>T</td>\n",
       "      <td>TCTGC</td>\n",
       "      <td>p.L287fs</td>\n",
       "      <td>exonic</td>\n",
       "      <td>frameshift insertion</td>\n",
       "      <td>0.4249</td>\n",
       "      <td>0.4608</td>\n",
       "      <td>1285</td>\n",
       "      <td>4226</td>\n",
       "      <td>606</td>\n",
       "      <td>5254</td>\n",
       "      <td>yes</td>\n",
       "      <td>bulk-NGS</td>\n",
       "      <td>0.1828</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>AML-01-001</td>\n",
       "      <td>WT1</td>\n",
       "      <td>11</td>\n",
       "      <td>32417913</td>\n",
       "      <td>C</td>\n",
       "      <td>CGTACAAGA</td>\n",
       "      <td>p.R380fs</td>\n",
       "      <td>exonic</td>\n",
       "      <td>frameshift insertion</td>\n",
       "      <td>0.0034</td>\n",
       "      <td>0.0015</td>\n",
       "      <td>10779</td>\n",
       "      <td>39</td>\n",
       "      <td>0</td>\n",
       "      <td>553</td>\n",
       "      <td>yes</td>\n",
       "      <td>bulk-NGS in paired sample</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>AML-01-001</td>\n",
       "      <td>WT1</td>\n",
       "      <td>11</td>\n",
       "      <td>32417943</td>\n",
       "      <td>C</td>\n",
       "      <td>CG</td>\n",
       "      <td>p.R370fs</td>\n",
       "      <td>exonic</td>\n",
       "      <td>frameshift insertion</td>\n",
       "      <td>0.8836</td>\n",
       "      <td>0.4865</td>\n",
       "      <td>568</td>\n",
       "      <td>9799</td>\n",
       "      <td>248</td>\n",
       "      <td>756</td>\n",
       "      <td>yes</td>\n",
       "      <td>bulk-NGS</td>\n",
       "      <td>0.236</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>AML-02-001</td>\n",
       "      <td>DNMT3A</td>\n",
       "      <td>2</td>\n",
       "      <td>25457242</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>p.R882H</td>\n",
       "      <td>exonic</td>\n",
       "      <td>nonsynonymous SNV</td>\n",
       "      <td>0.7603</td>\n",
       "      <td>0.4639</td>\n",
       "      <td>468</td>\n",
       "      <td>5861</td>\n",
       "      <td>169</td>\n",
       "      <td>1433</td>\n",
       "      <td>yes</td>\n",
       "      <td>bulk-NGS</td>\n",
       "      <td>0.4239</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    sample ID    gene chr      start ref allele  \\\n",
       "0  AML-01-001    FLT3  13   28608259          X   \n",
       "1  AML-01-001    NPM1   5  170837544          T   \n",
       "2  AML-01-001     WT1  11   32417913          C   \n",
       "3  AML-01-001     WT1  11   32417943          C   \n",
       "4  AML-02-001  DNMT3A   2   25457242          C   \n",
       "\n",
       "                                          alt allele   amino acid change  \\\n",
       "0  XAACTCTAAATTTTCTCTTGGAAACTCCCATTTGAGATCATATTCA...  p.Y599_D600delinsX   \n",
       "1                                              TCTGC            p.L287fs   \n",
       "2                                          CGTACAAGA            p.R380fs   \n",
       "3                                                 CG            p.R370fs   \n",
       "4                                                  T             p.R882H   \n",
       "\n",
       "  function       exonic function  mutated %  scDNA-seq VAF  \\\n",
       "0   exonic              stopgain     0.9352         0.7949   \n",
       "1   exonic  frameshift insertion     0.4249         0.4608   \n",
       "2   exonic  frameshift insertion     0.0034         0.0015   \n",
       "3   exonic  frameshift insertion     0.8836         0.4865   \n",
       "4   exonic     nonsynonymous SNV     0.7603         0.4639   \n",
       "\n",
       "   WT\\n(No. of cells)  Het\\n(No. of cells)  Homo\\n(No. of cells)  \\\n",
       "0                 239                 8444                  2190   \n",
       "1                1285                 4226                   606   \n",
       "2               10779                   39                     0   \n",
       "3                 568                 9799                   248   \n",
       "4                 468                 5861                   169   \n",
       "\n",
       "   Missing\\n(No. of cells) validated          validation method bulk VAF  \n",
       "0                      498       yes                   bulk-NGS   0.1186  \n",
       "1                     5254       yes                   bulk-NGS   0.1828  \n",
       "2                      553       yes  bulk-NGS in paired sample        0  \n",
       "3                      756       yes                   bulk-NGS    0.236  \n",
       "4                     1433       yes                   bulk-NGS   0.4239  "
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "AML_variants.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Next, we import the haploid trinucleotide-context-site-specific mutation rates from [Watson et al., 2020](https://www.science.org/doi/10.1126/science.aay9333)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "site_mutation_rates = pd.read_csv(\"../data/trinucleotide_mutation_rates_Watson_2020.csv\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Site</th>\n",
       "      <th>Mutation_rate</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>A[C&gt;A]A</td>\n",
       "      <td>1.332230e-09</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>A[C&gt;A]C</td>\n",
       "      <td>9.481170e-10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>A[C&gt;A]G</td>\n",
       "      <td>1.064910e-09</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>A[C&gt;A]T</td>\n",
       "      <td>7.585960e-10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>C[C&gt;A]A</td>\n",
       "      <td>1.538010e-09</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      Site  Mutation_rate\n",
       "0  A[C>A]A   1.332230e-09\n",
       "1  A[C>A]C   9.481170e-10\n",
       "2  A[C>A]G   1.064910e-09\n",
       "3  A[C>A]T   7.585960e-10\n",
       "4  C[C>A]A   1.538010e-09"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "site_mutation_rates.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Then we need the human genome asembly file (hg19), which needs to be downloaded from [here](https://www.gencodegenes.org/human/release_19.html) and saved to the same folder with name \"GRCh37.p13.genome.fa\"."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "hg19_fasta = Fasta(\"../data/GRCh37.p13.genome.fa\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "For each AML variant, check whether it is an SNV. If yes, then query the trinucleotide context from the fasta file based on the columns \"chr\" and \"start\" and determine its haploid trinucleotide-context-site-specific mutation rate."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "AML_variants[\"tri_context\"] = \"\"\n",
    "AML_variants[\"site_specific_mutation_rate\"] = 0.0\n",
    "for index, row in AML_variants.iterrows():\n",
    "\tif row[\"exonic function\"] == \"nonsynonymous SNV\":\n",
    "\t\tchrom = \"chr\" + row[\"chr\"]\n",
    "\t\tstart = row[\"start\"] - 2\n",
    "\t\tend = row[\"start\"] + 1\n",
    "\t\tsequence = hg19_fasta[chrom][start:end]\n",
    "\t\tp0 = sequence[0]\n",
    "\t\tp1 = sequence[1]\n",
    "\t\tp2 = sequence[2]\n",
    "\t\ttri_context = f\"{p0}[{p1}>{row['alt allele']}]{p2}\"\n",
    "\t\ttri_rate = site_mutation_rates[site_mutation_rates[\"Site\"] == tri_context][\"Mutation_rate\"].values[0]\n",
    "\t\tAML_variants.at[index, \"tri_context\"] = tri_context\n",
    "\t\tAML_variants.at[index, \"site_specific_mutation_rate\"] = tri_rate"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "For all other variants, we assign an average mutation rate of $2.7 Ã— 10^{âˆ’9}$ per bp per year."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "AML_variants[\"site_specific_mutation_rate\"] = AML_variants[\"site_specific_mutation_rate\"].replace(0.0, 2.7e-9)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Get all unique combinations of (gene, amino acid change, site_specific_mutation_rate):"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [],
   "source": [
    "AML_variants_unique = AML_variants.drop_duplicates(subset=[\"gene\", \"amino acid change\", \"ref allele\", \"alt allele\", \"site_specific_mutation_rate\"])\n",
    "AML_variants_unique = AML_variants_unique[[\"gene\", \"amino acid change\", \"ref allele\", \"alt allele\", \"exonic function\", \"site_specific_mutation_rate\"]]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene</th>\n",
       "      <th>amino acid change</th>\n",
       "      <th>ref allele</th>\n",
       "      <th>alt allele</th>\n",
       "      <th>exonic function</th>\n",
       "      <th>site_specific_mutation_rate</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>50</th>\n",
       "      <td>TET2</td>\n",
       "      <td>p.P1215L</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>nonsynonymous SNV</td>\n",
       "      <td>4.379090e-09</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>98</th>\n",
       "      <td>TET2</td>\n",
       "      <td>p.E375X</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>stopgain</td>\n",
       "      <td>2.700000e-09</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>318</th>\n",
       "      <td>TET2</td>\n",
       "      <td>splicing</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>.</td>\n",
       "      <td>2.700000e-09</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>319</th>\n",
       "      <td>TET2</td>\n",
       "      <td>p.H1889Y</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>nonsynonymous SNV</td>\n",
       "      <td>9.919690e-10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>323</th>\n",
       "      <td>TET2</td>\n",
       "      <td>p.G1882R</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>nonsynonymous SNV</td>\n",
       "      <td>2.646110e-09</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>360</th>\n",
       "      <td>TET2</td>\n",
       "      <td>p.I274fs</td>\n",
       "      <td>TC</td>\n",
       "      <td>T</td>\n",
       "      <td>frameshift deletion</td>\n",
       "      <td>2.700000e-09</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>384</th>\n",
       "      <td>TET2</td>\n",
       "      <td>p.Q1009X</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>stopgain</td>\n",
       "      <td>2.700000e-09</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>386</th>\n",
       "      <td>TET2</td>\n",
       "      <td>p.E1534K</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>nonsynonymous SNV</td>\n",
       "      <td>4.379090e-09</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>401</th>\n",
       "      <td>TET2</td>\n",
       "      <td>p.S1267X</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>stopgain</td>\n",
       "      <td>2.700000e-09</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>410</th>\n",
       "      <td>TET2</td>\n",
       "      <td>p.Q866X</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>stopgain</td>\n",
       "      <td>2.700000e-09</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>417</th>\n",
       "      <td>TET2</td>\n",
       "      <td>p.Q1567X</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>stopgain</td>\n",
       "      <td>2.700000e-09</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>422</th>\n",
       "      <td>TET2</td>\n",
       "      <td>p.P1638fs</td>\n",
       "      <td>C</td>\n",
       "      <td>CA</td>\n",
       "      <td>frameshift insertion</td>\n",
       "      <td>2.700000e-09</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>426</th>\n",
       "      <td>TET2</td>\n",
       "      <td>p.K1329X</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>stopgain</td>\n",
       "      <td>2.700000e-09</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>431</th>\n",
       "      <td>TET2</td>\n",
       "      <td>p.G1296R</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>nonsynonymous SNV</td>\n",
       "      <td>4.559440e-10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>435</th>\n",
       "      <td>TET2</td>\n",
       "      <td>p.N926D</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>nonsynonymous SNV</td>\n",
       "      <td>4.958240e-10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>465</th>\n",
       "      <td>TET2</td>\n",
       "      <td>p.K228R</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>nonsynonymous SNV</td>\n",
       "      <td>4.958240e-10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>476</th>\n",
       "      <td>TET2</td>\n",
       "      <td>p.Q1563X</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>stopgain</td>\n",
       "      <td>2.700000e-09</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>501</th>\n",
       "      <td>TET2</td>\n",
       "      <td>p.N483fs</td>\n",
       "      <td>AT</td>\n",
       "      <td>A</td>\n",
       "      <td>frameshift deletion</td>\n",
       "      <td>2.700000e-09</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>502</th>\n",
       "      <td>TET2</td>\n",
       "      <td>p.R1387fs</td>\n",
       "      <td>GT</td>\n",
       "      <td>G</td>\n",
       "      <td>frameshift deletion</td>\n",
       "      <td>2.700000e-09</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>517</th>\n",
       "      <td>TET2</td>\n",
       "      <td>p.L1046fs</td>\n",
       "      <td>T</td>\n",
       "      <td>TA</td>\n",
       "      <td>frameshift insertion</td>\n",
       "      <td>2.700000e-09</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>518</th>\n",
       "      <td>TET2</td>\n",
       "      <td>p.R1237X</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>stopgain</td>\n",
       "      <td>2.700000e-09</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>524</th>\n",
       "      <td>TET2</td>\n",
       "      <td>p.E1255X</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>stopgain</td>\n",
       "      <td>2.700000e-09</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>525</th>\n",
       "      <td>TET2</td>\n",
       "      <td>p.G1382D</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>nonsynonymous SNV</td>\n",
       "      <td>3.175730e-09</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>541</th>\n",
       "      <td>TET2</td>\n",
       "      <td>p.H717fs</td>\n",
       "      <td>C</td>\n",
       "      <td>CAA</td>\n",
       "      <td>frameshift insertion</td>\n",
       "      <td>2.700000e-09</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     gene amino acid change ref allele alt allele       exonic function  \\\n",
       "50   TET2          p.P1215L          C          T     nonsynonymous SNV   \n",
       "98   TET2           p.E375X          G          T              stopgain   \n",
       "318  TET2          splicing          G          A                     .   \n",
       "319  TET2          p.H1889Y          C          T     nonsynonymous SNV   \n",
       "323  TET2          p.G1882R          G          A     nonsynonymous SNV   \n",
       "360  TET2          p.I274fs         TC          T   frameshift deletion   \n",
       "384  TET2          p.Q1009X          C          T              stopgain   \n",
       "386  TET2          p.E1534K          G          A     nonsynonymous SNV   \n",
       "401  TET2          p.S1267X          C          A              stopgain   \n",
       "410  TET2           p.Q866X          C          T              stopgain   \n",
       "417  TET2          p.Q1567X          C          T              stopgain   \n",
       "422  TET2         p.P1638fs          C         CA  frameshift insertion   \n",
       "426  TET2          p.K1329X          A          T              stopgain   \n",
       "431  TET2          p.G1296R          G          C     nonsynonymous SNV   \n",
       "435  TET2           p.N926D          A          G     nonsynonymous SNV   \n",
       "465  TET2           p.K228R          A          G     nonsynonymous SNV   \n",
       "476  TET2          p.Q1563X          C          T              stopgain   \n",
       "501  TET2          p.N483fs         AT          A   frameshift deletion   \n",
       "502  TET2         p.R1387fs         GT          G   frameshift deletion   \n",
       "517  TET2         p.L1046fs          T         TA  frameshift insertion   \n",
       "518  TET2          p.R1237X          C          T              stopgain   \n",
       "524  TET2          p.E1255X          G          T              stopgain   \n",
       "525  TET2          p.G1382D          G          A     nonsynonymous SNV   \n",
       "541  TET2          p.H717fs          C        CAA  frameshift insertion   \n",
       "\n",
       "     site_specific_mutation_rate  \n",
       "50                  4.379090e-09  \n",
       "98                  2.700000e-09  \n",
       "318                 2.700000e-09  \n",
       "319                 9.919690e-10  \n",
       "323                 2.646110e-09  \n",
       "360                 2.700000e-09  \n",
       "384                 2.700000e-09  \n",
       "386                 4.379090e-09  \n",
       "401                 2.700000e-09  \n",
       "410                 2.700000e-09  \n",
       "417                 2.700000e-09  \n",
       "422                 2.700000e-09  \n",
       "426                 2.700000e-09  \n",
       "431                 4.559440e-10  \n",
       "435                 4.958240e-10  \n",
       "465                 4.958240e-10  \n",
       "476                 2.700000e-09  \n",
       "501                 2.700000e-09  \n",
       "502                 2.700000e-09  \n",
       "517                 2.700000e-09  \n",
       "518                 2.700000e-09  \n",
       "524                 2.700000e-09  \n",
       "525                 3.175730e-09  \n",
       "541                 2.700000e-09  "
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "AML_variants_unique.query(\"gene == 'TET2'\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "For each gene, sum up the site-specific mutation rates."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [],
   "source": [
    "AML_variants_sum = AML_variants_unique.groupby(\"gene\").agg({\"site_specific_mutation_rate\": \"sum\"}).reset_index()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene</th>\n",
       "      <th>site_specific_mutation_rate</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ASXL1</td>\n",
       "      <td>2.430000e-08</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>BCOR</td>\n",
       "      <td>1.350000e-08</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>CBL</td>\n",
       "      <td>1.630361e-08</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>CSF3R</td>\n",
       "      <td>2.700000e-09</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>DNMT3A</td>\n",
       "      <td>1.137878e-07</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>ETV6</td>\n",
       "      <td>2.700000e-09</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>EZH2</td>\n",
       "      <td>5.053607e-08</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>FLT3</td>\n",
       "      <td>1.297272e-07</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>GATA2</td>\n",
       "      <td>6.238082e-09</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>IDH1</td>\n",
       "      <td>4.526963e-08</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>IDH2</td>\n",
       "      <td>1.853029e-08</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>JAK2</td>\n",
       "      <td>1.332230e-09</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>KIT</td>\n",
       "      <td>4.533170e-10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>KRAS</td>\n",
       "      <td>4.925357e-08</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>MPL</td>\n",
       "      <td>1.415120e-08</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>MYC</td>\n",
       "      <td>1.577025e-08</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>NPM1</td>\n",
       "      <td>1.350000e-08</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>NRAS</td>\n",
       "      <td>2.034380e-08</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>PHF6</td>\n",
       "      <td>1.470440e-08</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>PPM1D</td>\n",
       "      <td>8.100000e-09</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>PTPN11</td>\n",
       "      <td>1.854313e-08</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>RUNX1</td>\n",
       "      <td>1.280074e-07</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>SETBP1</td>\n",
       "      <td>1.200440e-08</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>SF3B1</td>\n",
       "      <td>2.241882e-09</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>SMC3</td>\n",
       "      <td>4.832630e-10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>SRSF2</td>\n",
       "      <td>4.372184e-09</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>STAG2</td>\n",
       "      <td>8.100000e-09</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>TET2</td>\n",
       "      <td>6.021958e-08</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>TP53</td>\n",
       "      <td>5.109518e-08</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>U2AF1</td>\n",
       "      <td>2.698110e-09</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>WT1</td>\n",
       "      <td>9.685030e-08</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      gene  site_specific_mutation_rate\n",
       "0    ASXL1                 2.430000e-08\n",
       "1     BCOR                 1.350000e-08\n",
       "2      CBL                 1.630361e-08\n",
       "3    CSF3R                 2.700000e-09\n",
       "4   DNMT3A                 1.137878e-07\n",
       "5     ETV6                 2.700000e-09\n",
       "6     EZH2                 5.053607e-08\n",
       "7     FLT3                 1.297272e-07\n",
       "8    GATA2                 6.238082e-09\n",
       "9     IDH1                 4.526963e-08\n",
       "10    IDH2                 1.853029e-08\n",
       "11    JAK2                 1.332230e-09\n",
       "12     KIT                 4.533170e-10\n",
       "13    KRAS                 4.925357e-08\n",
       "14     MPL                 1.415120e-08\n",
       "15     MYC                 1.577025e-08\n",
       "16    NPM1                 1.350000e-08\n",
       "17    NRAS                 2.034380e-08\n",
       "18    PHF6                 1.470440e-08\n",
       "19   PPM1D                 8.100000e-09\n",
       "20  PTPN11                 1.854313e-08\n",
       "21   RUNX1                 1.280074e-07\n",
       "22  SETBP1                 1.200440e-08\n",
       "23   SF3B1                 2.241882e-09\n",
       "24    SMC3                 4.832630e-10\n",
       "25   SRSF2                 4.372184e-09\n",
       "26   STAG2                 8.100000e-09\n",
       "27    TET2                 6.021958e-08\n",
       "28    TP53                 5.109518e-08\n",
       "29   U2AF1                 2.698110e-09\n",
       "30     WT1                 9.685030e-08"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "AML_variants_sum"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# save the AML variants to a file\n",
    "AML_variants_sum.to_csv(\"../data/AML_mutation_rates.csv\", index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [],
   "source": [
    "AML_variants_sum_dict = dict(zip(AML_variants_sum[\"gene\"], AML_variants_sum[\"site_specific_mutation_rate\"]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'ASXL1': 2.4300000000000003e-08,\n",
       " 'BCOR': 1.3500000000000002e-08,\n",
       " 'CBL': 1.630361e-08,\n",
       " 'CSF3R': 2.7e-09,\n",
       " 'DNMT3A': 1.13787835e-07,\n",
       " 'ETV6': 2.7e-09,\n",
       " 'EZH2': 5.0536069e-08,\n",
       " 'FLT3': 1.29727249e-07,\n",
       " 'GATA2': 6.238082000000001e-09,\n",
       " 'IDH1': 4.526963e-08,\n",
       " 'IDH2': 1.853029e-08,\n",
       " 'JAK2': 1.33223e-09,\n",
       " 'KIT': 4.53317e-10,\n",
       " 'KRAS': 4.9253573e-08,\n",
       " 'MPL': 1.41512e-08,\n",
       " 'MYC': 1.5770254e-08,\n",
       " 'NPM1': 1.3500000000000002e-08,\n",
       " 'NRAS': 2.0343801e-08,\n",
       " 'PHF6': 1.4704399999999999e-08,\n",
       " 'PPM1D': 8.1e-09,\n",
       " 'PTPN11': 1.8543133e-08,\n",
       " 'RUNX1': 1.2800741e-07,\n",
       " 'SETBP1': 1.20044e-08,\n",
       " 'SF3B1': 2.241882e-09,\n",
       " 'SMC3': 4.83263e-10,\n",
       " 'SRSF2': 4.372184e-09,\n",
       " 'STAG2': 8.1e-09,\n",
       " 'TET2': 6.021958100000001e-08,\n",
       " 'TP53': 5.1095175e-08,\n",
       " 'U2AF1': 2.69811e-09,\n",
       " 'WT1': 9.68503e-08}"
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "AML_variants_sum_dict"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### 4. Prepare trees"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now we are ready to parse the trees. We loop through the tree dataframe. For each patient, create a *TumorTree* object by first creating a root node and then adding all the other nodes as *Subclone*."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [],
   "source": [
    "trees = []\n",
    "\n",
    "for patient_id in AML_tree_df_combined[\"Patient_ID\"].unique():\n",
    "\tpatient_df = AML_tree_df_combined[AML_tree_df_combined[\"Patient_ID\"] == patient_id]\n",
    "\tsampling_time = patient_df[\"Sampling_Time\"].iloc[0]\n",
    "\ttree_id = patient_df[\"Tree_ID\"].iloc[0]\n",
    "\troot_node = Subclone(node_id=0, mutation_ids=[], seq_cell_number=C_0)  # pyright: ignore\n",
    "\tnode_dict = {0: root_node}\n",
    "\n",
    "\t# Loop through rows of patient_df\n",
    "\tfor index, row in patient_df.iterrows():\n",
    "\t\tnew_node = Subclone(\n",
    "\t\t\tnode_id=row[\"Node_ID\"],\n",
    "\t\t\tmutation_ids=[row[\"Mutation_ID\"]],\n",
    "\t\t\tseq_cell_number=row[\"Sequenced_Cells\"],\n",
    "\t\t\tparent=node_dict[row[\"Parent_ID\"]],\n",
    "\t\t)\n",
    "\t\tnode_dict[row[\"Node_ID\"]] = new_node\n",
    "\t\n",
    "\ttree = TumorTree(\n",
    "\t\tpatient_id=patient_id,\n",
    "\t\ttree_id=tree_id,\n",
    "\t\tsampling_time=sampling_time,\n",
    "\t\troot=root_node,\n",
    "\t\ttumor_size=patient_df[\"Total_Cells\"].iloc[0],\n",
    "\t)\n",
    "\ttrees.append(tree)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Finally, we construct the *TumorTreeCohort* object."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [],
   "source": [
    "name = \"AML_cohort_Morita_2020\"\n",
    "n_mutations = AML_tree_df_combined[\"Mutation_ID\"].nunique()\n",
    "N_patients = AML_tree_df_combined[\"Patient_ID\"].nunique()\n",
    "N_trees = N_patients\n",
    "mutation_labels = AML_tree_df_combined[[\"Mutation_ID\", \"Mutation\"]].drop_duplicates().sort_values(\"Mutation_ID\")[\"Mutation\"].values\n",
    "patient_labels = AML_tree_df_combined[[\"Patient_ID\", \"Patient\"]].drop_duplicates().sort_values(\"Patient_ID\")[\"Patient\"].values\n",
    "tree_labels = patient_labels\n",
    "mu_vec = np.array([AML_variants_sum_dict.get(mutation, 0.0) for mutation in mutation_labels])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [],
   "source": [
    "AML_cohort = TumorTreeCohort(\n",
    "\tname=name,\n",
    "\ttrees=trees,\n",
    "\tn_mutations=n_mutations,\n",
    "\tN_patients=N_patients,\n",
    "\tN_trees=N_trees,\n",
    "\tmu_vec=mu_vec,\n",
    "\tcommon_beta=common_beta,\n",
    "\tC_0=C_0,\n",
    "\tC_sampling=C_sampling,\n",
    "\tt_max=t_max,\n",
    "\tmutation_labels=mutation_labels,\n",
    "\tpatient_labels=patient_labels,\n",
    "\ttree_labels=tree_labels,\n",
    "\tlifetime_risk=lifetime_risk,\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The followings are two i/o functions:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "save_cohort_to_json(AML_cohort, \"../data/AML_cohort_Morita_2020.json\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [],
   "source": [
    "AML_cohort = load_cohort_from_json(\"../data/AML_cohort_Morita_2020.json\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "There are two print/plot functions:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Tumor tree 0 of patient                 0 at time 48\n",
      " - Total number of tumor cells: 9.4900E+11\n",
      "(Node ID: 0) \n",
      " - Mutations: [] \n",
      " - Cell number: 1.0000E+05 \n",
      " - Sequence cell number: 1.0000E+05 \n",
      "â””â”€â”€ (Node ID: 1) \n",
      "     - Mutations: [0] \n",
      "     - Cell number: 3.3422E+10 \n",
      "     - Sequence cell number: 3.9000E+02 \n",
      "    â””â”€â”€ (Node ID: 2) \n",
      "         - Mutations: [2] \n",
      "         - Cell number: 2.8948E+11 \n",
      "         - Sequence cell number: 3.3780E+03 \n",
      "        â””â”€â”€ (Node ID: 3) \n",
      "             - Mutations: [1] \n",
      "             - Cell number: 6.2610E+11 \n",
      "             - Sequence cell number: 7.3060E+03 \n"
     ]
    }
   ],
   "source": [
    "print(AML_cohort.trees[0])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning: node 's_1', graph 'AML-01-001' size too small for label\n"
     ]
    },
    {
     "data": {
      "image/svg+xml": [
       "<?xml version=\"1.0\" encoding=\"UTF-8\" standalone=\"no\"?>\n",
       "<!DOCTYPE svg PUBLIC \"-//W3C//DTD SVG 1.1//EN\"\n",
       " \"http://www.w3.org/Graphics/SVG/1.1/DTD/svg11.dtd\">\n",
       "<!-- Generated by graphviz version 12.2.1 (20241206.2353)\n",
       " -->\n",
       "<!-- Title: AML&#45;01&#45;001 Pages: 1 -->\n",
       "<svg width=\"687pt\" height=\"720pt\"\n",
       " viewBox=\"0.00 0.00 687.16 720.00\" xmlns=\"http://www.w3.org/2000/svg\" xmlns:xlink=\"http://www.w3.org/1999/xlink\">\n",
       "<g id=\"graph0\" class=\"graph\" transform=\"scale(0.899811 0.899811) rotate(0) translate(4 796.17)\">\n",
       "<title>AML&#45;01&#45;001</title>\n",
       "<polygon fill=\"white\" stroke=\"none\" points=\"-4,4 -4,-796.17 759.67,-796.17 759.67,4 -4,4\"/>\n",
       "<text text-anchor=\"middle\" x=\"377.83\" y=\"-761.57\" font-family=\"Helvetica,sans-Serif\" font-size=\"28.00\">Tree: AML&#45;01&#45;001</text>\n",
       "<!-- 0 -->\n",
       "<g id=\"node1\" class=\"node\">\n",
       "<title>0</title>\n",
       "<ellipse fill=\"none\" stroke=\"dimgray\" stroke-width=\"5\" cx=\"198.5\" cy=\"-733\" rx=\"47.38\" ry=\"23.16\"/>\n",
       "<text text-anchor=\"middle\" x=\"198.5\" y=\"-722.58\" font-family=\"Helvetica,sans-Serif\" font-size=\"24.00\" fill=\"dimgray\">Root</text>\n",
       "</g>\n",
       "<!-- 1 -->\n",
       "<g id=\"node2\" class=\"node\">\n",
       "<title>1</title>\n",
       "<ellipse fill=\"none\" stroke=\"dimgray\" stroke-width=\"5\" cx=\"198.5\" cy=\"-650\" rx=\"50.03\" ry=\"23.16\"/>\n",
       "<text text-anchor=\"middle\" x=\"198.5\" y=\"-639.58\" font-family=\"Helvetica,sans-Serif\" font-size=\"24.00\" fill=\"dimgray\">FLT3</text>\n",
       "</g>\n",
       "<!-- 0&#45;&gt;1 -->\n",
       "<g id=\"edge1\" class=\"edge\">\n",
       "<title>0&#45;&gt;1</title>\n",
       "<path fill=\"none\" stroke=\"dimgray\" stroke-width=\"5\" d=\"M198.5,-707.7C198.5,-702.62 198.5,-697.13 198.5,-691.64\"/>\n",
       "<polygon fill=\"dimgray\" stroke=\"dimgray\" stroke-width=\"5\" points=\"202.87,-691.8 198.5,-681.8 194.12,-691.8 202.87,-691.8\"/>\n",
       "</g>\n",
       "<!-- s_1 -->\n",
       "<g id=\"node3\" class=\"node\">\n",
       "<title>s_1</title>\n",
       "<ellipse fill=\"#6ca6cd\" stroke=\"#6ca6cd\" stroke-width=\"5\" cx=\"45.5\" cy=\"-564\" rx=\"27.02\" ry=\"27.02\"/>\n",
       "<text text-anchor=\"middle\" x=\"45.5\" y=\"-565.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"16.00\">3.3422E+10</text>\n",
       "<text text-anchor=\"middle\" x=\"45.5\" y=\"-548.8\" font-family=\"Helvetica,sans-Serif\" font-size=\"16.00\">(3.52%)</text>\n",
       "</g>\n",
       "<!-- 1&#45;&gt;s_1 -->\n",
       "<g id=\"edge2\" class=\"edge\">\n",
       "<title>1&#45;&gt;s_1</title>\n",
       "<path fill=\"none\" stroke=\"#6ca6cd\" stroke-width=\"5\" stroke-dasharray=\"5,2\" d=\"M164.92,-630.57C136.6,-615.02 96.66,-593.09 70.93,-578.96\"/>\n",
       "</g>\n",
       "<!-- 2 -->\n",
       "<g id=\"node4\" class=\"node\">\n",
       "<title>2</title>\n",
       "<ellipse fill=\"none\" stroke=\"dimgray\" stroke-width=\"5\" cx=\"351.5\" cy=\"-564\" rx=\"47.38\" ry=\"23.16\"/>\n",
       "<text text-anchor=\"middle\" x=\"351.5\" y=\"-553.58\" font-family=\"Helvetica,sans-Serif\" font-size=\"24.00\" fill=\"dimgray\">WT1</text>\n",
       "</g>\n",
       "<!-- 1&#45;&gt;2 -->\n",
       "<g id=\"edge3\" class=\"edge\">\n",
       "<title>1&#45;&gt;2</title>\n",
       "<path fill=\"none\" stroke=\"dimgray\" stroke-width=\"5\" d=\"M232.07,-630.57C253.34,-618.89 281.16,-603.62 304.6,-590.75\"/>\n",
       "<polygon fill=\"dimgray\" stroke=\"dimgray\" stroke-width=\"5\" points=\"306.4,-594.75 313.06,-586.1 302.19,-587.08 306.4,-594.75\"/>\n",
       "</g>\n",
       "<!-- s_2 -->\n",
       "<g id=\"node5\" class=\"node\">\n",
       "<title>s_2</title>\n",
       "<ellipse fill=\"sandybrown\" stroke=\"sandybrown\" stroke-width=\"5\" cx=\"144.5\" cy=\"-421\" rx=\"79.53\" ry=\"79.53\"/>\n",
       "<text text-anchor=\"middle\" x=\"144.5\" y=\"-422.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"16.00\">2.8948E+11</text>\n",
       "<text text-anchor=\"middle\" x=\"144.5\" y=\"-405.8\" font-family=\"Helvetica,sans-Serif\" font-size=\"16.00\">(30.50%)</text>\n",
       "</g>\n",
       "<!-- 2&#45;&gt;s_2 -->\n",
       "<g id=\"edge4\" class=\"edge\">\n",
       "<title>2&#45;&gt;s_2</title>\n",
       "<path fill=\"none\" stroke=\"sandybrown\" stroke-width=\"5\" stroke-dasharray=\"5,2\" d=\"M322.71,-543.39C294.05,-523.87 249.02,-493.2 211.4,-467.57\"/>\n",
       "</g>\n",
       "<!-- 3 -->\n",
       "<g id=\"node6\" class=\"node\">\n",
       "<title>3</title>\n",
       "<ellipse fill=\"none\" stroke=\"dimgray\" stroke-width=\"5\" cx=\"558.5\" cy=\"-421\" rx=\"58.51\" ry=\"23.16\"/>\n",
       "<text text-anchor=\"middle\" x=\"558.5\" y=\"-410.57\" font-family=\"Helvetica,sans-Serif\" font-size=\"24.00\" fill=\"dimgray\">NPM1</text>\n",
       "</g>\n",
       "<!-- 2&#45;&gt;3 -->\n",
       "<g id=\"edge5\" class=\"edge\">\n",
       "<title>2&#45;&gt;3</title>\n",
       "<path fill=\"none\" stroke=\"dimgray\" stroke-width=\"5\" d=\"M380.28,-543.39C415.04,-519.72 473.87,-479.65 514.37,-452.06\"/>\n",
       "<polygon fill=\"dimgray\" stroke=\"dimgray\" stroke-width=\"5\" points=\"516.64,-455.8 522.44,-446.56 511.72,-448.57 516.64,-455.8\"/>\n",
       "</g>\n",
       "<!-- s_3 -->\n",
       "<g id=\"node7\" class=\"node\">\n",
       "<title>s_3</title>\n",
       "<ellipse fill=\"#96cdcd\" stroke=\"#96cdcd\" stroke-width=\"5\" cx=\"558.5\" cy=\"-189\" rx=\"116.96\" ry=\"116.96\"/>\n",
       "<text text-anchor=\"middle\" x=\"558.5\" y=\"-190.3\" font-family=\"Helvetica,sans-Serif\" font-size=\"16.00\">6.2610E+11</text>\n",
       "<text text-anchor=\"middle\" x=\"558.5\" y=\"-173.8\" font-family=\"Helvetica,sans-Serif\" font-size=\"16.00\">(65.97%)</text>\n",
       "</g>\n",
       "<!-- 3&#45;&gt;s_3 -->\n",
       "<g id=\"edge6\" class=\"edge\">\n",
       "<title>3&#45;&gt;s_3</title>\n",
       "<path fill=\"none\" stroke=\"#96cdcd\" stroke-width=\"5\" stroke-dasharray=\"5,2\" d=\"M558.5,-395.39C558.5,-374 558.5,-341.36 558.5,-308.35\"/>\n",
       "</g>\n",
       "<!-- sampling_time -->\n",
       "<g id=\"node8\" class=\"node\">\n",
       "<title>sampling_time</title>\n",
       "<polygon fill=\"none\" stroke=\"dimgray\" stroke-width=\"5\" points=\"663.25,-36 453.75,-36 453.75,0 663.25,0 663.25,-36\"/>\n",
       "<text text-anchor=\"middle\" x=\"558.5\" y=\"-7.58\" font-family=\"Helvetica,sans-Serif\" font-size=\"24.00\" fill=\"dimgray\">Sampling time: 48</text>\n",
       "</g>\n",
       "</g>\n",
       "</svg>\n"
      ],
      "text/plain": [
       "<graphviz.sources.Source at 0x3046dd590>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "i = 0\n",
    "plot_tree(AML_cohort, i)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "FiTree",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.9"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
