{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Estimation of AML Mutation Rates"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Import necessary libraries:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "import fitree\n",
    "import pandas as pd\n",
    "from pyfaidx import Fasta"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Load the AML tree data:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "AML_cohort = fitree.load_cohort_from_json(\"../data/AML_morita/AML_cohort_Morita_2020.json\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "All 31 genes:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['FLT3', 'NPM1', 'WT1', 'DNMT3A', 'KRAS', 'NRAS', 'RUNX1', 'IDH1', 'IDH2', 'PTPN11', 'SRSF2', 'ASXL1', 'BCOR', 'STAG2', 'TP53', 'U2AF1', 'SF3B1', 'TET2', 'CSF3R', 'JAK2', 'GATA2', 'EZH2', 'PPM1D', 'SETBP1', 'KIT', 'CBL', 'PHF6', 'MYC', 'ETV6', 'MPL', 'SMC3']\n"
     ]
    }
   ],
   "source": [
    "AML_genes = AML_cohort.mutation_labels\n",
    "print(AML_genes)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Load the variant information:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "AML_variants = pd.read_csv(\"../data/AML_morita/AML_variants.csv\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Filter for those that have been validated:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "AML_variants = AML_variants[AML_variants[\"validated\"] == \"yes\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample ID</th>\n",
       "      <th>gene</th>\n",
       "      <th>chr</th>\n",
       "      <th>start</th>\n",
       "      <th>ref allele</th>\n",
       "      <th>alt allele</th>\n",
       "      <th>amino acid change</th>\n",
       "      <th>function</th>\n",
       "      <th>exonic function</th>\n",
       "      <th>mutated %</th>\n",
       "      <th>scDNA-seq VAF</th>\n",
       "      <th>WT\\n(No. of cells)</th>\n",
       "      <th>Het\\n(No. of cells)</th>\n",
       "      <th>Homo\\n(No. of cells)</th>\n",
       "      <th>Missing\\n(No. of cells)</th>\n",
       "      <th>validated</th>\n",
       "      <th>validation method</th>\n",
       "      <th>bulk VAF</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AML-01-001</td>\n",
       "      <td>FLT3</td>\n",
       "      <td>13</td>\n",
       "      <td>28608259</td>\n",
       "      <td>X</td>\n",
       "      <td>XAACTCTAAATTTTCTCTTGGAAACTCCCATTTGAGATCATATTCA...</td>\n",
       "      <td>p.Y599_D600delinsX</td>\n",
       "      <td>exonic</td>\n",
       "      <td>stopgain</td>\n",
       "      <td>0.935200</td>\n",
       "      <td>0.7949</td>\n",
       "      <td>239</td>\n",
       "      <td>8444</td>\n",
       "      <td>2190</td>\n",
       "      <td>498</td>\n",
       "      <td>yes</td>\n",
       "      <td>bulk-NGS</td>\n",
       "      <td>0.1186</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>AML-01-001</td>\n",
       "      <td>NPM1</td>\n",
       "      <td>5</td>\n",
       "      <td>170837544</td>\n",
       "      <td>T</td>\n",
       "      <td>TCTGC</td>\n",
       "      <td>p.L287fs</td>\n",
       "      <td>exonic</td>\n",
       "      <td>frameshift insertion</td>\n",
       "      <td>0.424900</td>\n",
       "      <td>0.4608</td>\n",
       "      <td>1285</td>\n",
       "      <td>4226</td>\n",
       "      <td>606</td>\n",
       "      <td>5254</td>\n",
       "      <td>yes</td>\n",
       "      <td>bulk-NGS</td>\n",
       "      <td>0.1828</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>AML-01-001</td>\n",
       "      <td>WT1</td>\n",
       "      <td>11</td>\n",
       "      <td>32417913</td>\n",
       "      <td>C</td>\n",
       "      <td>CGTACAAGA</td>\n",
       "      <td>p.R380fs</td>\n",
       "      <td>exonic</td>\n",
       "      <td>frameshift insertion</td>\n",
       "      <td>0.003400</td>\n",
       "      <td>0.0015</td>\n",
       "      <td>10779</td>\n",
       "      <td>39</td>\n",
       "      <td>0</td>\n",
       "      <td>553</td>\n",
       "      <td>yes</td>\n",
       "      <td>bulk-NGS in paired sample</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>AML-01-001</td>\n",
       "      <td>WT1</td>\n",
       "      <td>11</td>\n",
       "      <td>32417943</td>\n",
       "      <td>C</td>\n",
       "      <td>CG</td>\n",
       "      <td>p.R370fs</td>\n",
       "      <td>exonic</td>\n",
       "      <td>frameshift insertion</td>\n",
       "      <td>0.883600</td>\n",
       "      <td>0.4865</td>\n",
       "      <td>568</td>\n",
       "      <td>9799</td>\n",
       "      <td>248</td>\n",
       "      <td>756</td>\n",
       "      <td>yes</td>\n",
       "      <td>bulk-NGS</td>\n",
       "      <td>0.236</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>AML-02-001</td>\n",
       "      <td>DNMT3A</td>\n",
       "      <td>2</td>\n",
       "      <td>25457242</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>p.R882H</td>\n",
       "      <td>exonic</td>\n",
       "      <td>nonsynonymous SNV</td>\n",
       "      <td>0.760300</td>\n",
       "      <td>0.4639</td>\n",
       "      <td>468</td>\n",
       "      <td>5861</td>\n",
       "      <td>169</td>\n",
       "      <td>1433</td>\n",
       "      <td>yes</td>\n",
       "      <td>bulk-NGS</td>\n",
       "      <td>0.4239</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>538</th>\n",
       "      <td>AML-122-001</td>\n",
       "      <td>SF3B1</td>\n",
       "      <td>2</td>\n",
       "      <td>198267359</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>p.K666N</td>\n",
       "      <td>exonic</td>\n",
       "      <td>nonsynonymous SNV</td>\n",
       "      <td>0.903884</td>\n",
       "      <td>0.5040</td>\n",
       "      <td>197</td>\n",
       "      <td>3497</td>\n",
       "      <td>133</td>\n",
       "      <td>189</td>\n",
       "      <td>yes</td>\n",
       "      <td>clinical sequencing</td>\n",
       "      <td>not available</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>539</th>\n",
       "      <td>AML-123-001</td>\n",
       "      <td>DNMT3A</td>\n",
       "      <td>2</td>\n",
       "      <td>25457243</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>p.R882C</td>\n",
       "      <td>exonic</td>\n",
       "      <td>nonsynonymous SNV</td>\n",
       "      <td>0.631245</td>\n",
       "      <td>0.3954</td>\n",
       "      <td>643</td>\n",
       "      <td>1495</td>\n",
       "      <td>97</td>\n",
       "      <td>287</td>\n",
       "      <td>yes</td>\n",
       "      <td>clinical sequencing</td>\n",
       "      <td>not available</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>540</th>\n",
       "      <td>AML-123-001</td>\n",
       "      <td>RUNX1</td>\n",
       "      <td>21</td>\n",
       "      <td>36231774</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>p.R204X</td>\n",
       "      <td>exonic</td>\n",
       "      <td>stopgain</td>\n",
       "      <td>0.420698</td>\n",
       "      <td>0.2909</td>\n",
       "      <td>1061</td>\n",
       "      <td>1031</td>\n",
       "      <td>30</td>\n",
       "      <td>400</td>\n",
       "      <td>yes</td>\n",
       "      <td>clinical sequencing</td>\n",
       "      <td>not available</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>541</th>\n",
       "      <td>AML-123-001</td>\n",
       "      <td>TET2</td>\n",
       "      <td>4</td>\n",
       "      <td>106157248</td>\n",
       "      <td>C</td>\n",
       "      <td>CAA</td>\n",
       "      <td>p.H717fs</td>\n",
       "      <td>exonic</td>\n",
       "      <td>frameshift insertion</td>\n",
       "      <td>0.706186</td>\n",
       "      <td>0.3574</td>\n",
       "      <td>620</td>\n",
       "      <td>1751</td>\n",
       "      <td>30</td>\n",
       "      <td>121</td>\n",
       "      <td>yes</td>\n",
       "      <td>clinical sequencing</td>\n",
       "      <td>not available</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>542</th>\n",
       "      <td>AML-123-001</td>\n",
       "      <td>U2AF1</td>\n",
       "      <td>21</td>\n",
       "      <td>44524456</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>p.S34F</td>\n",
       "      <td>exonic</td>\n",
       "      <td>nonsynonymous SNV</td>\n",
       "      <td>0.649881</td>\n",
       "      <td>0.3488</td>\n",
       "      <td>734</td>\n",
       "      <td>1499</td>\n",
       "      <td>140</td>\n",
       "      <td>149</td>\n",
       "      <td>yes</td>\n",
       "      <td>clinical sequencing</td>\n",
       "      <td>not available</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>530 rows × 18 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       sample ID    gene chr      start ref allele  \\\n",
       "0     AML-01-001    FLT3  13   28608259          X   \n",
       "1     AML-01-001    NPM1   5  170837544          T   \n",
       "2     AML-01-001     WT1  11   32417913          C   \n",
       "3     AML-01-001     WT1  11   32417943          C   \n",
       "4     AML-02-001  DNMT3A   2   25457242          C   \n",
       "..           ...     ...  ..        ...        ...   \n",
       "538  AML-122-001   SF3B1   2  198267359          C   \n",
       "539  AML-123-001  DNMT3A   2   25457243          G   \n",
       "540  AML-123-001   RUNX1  21   36231774          G   \n",
       "541  AML-123-001    TET2   4  106157248          C   \n",
       "542  AML-123-001   U2AF1  21   44524456          G   \n",
       "\n",
       "                                            alt allele   amino acid change  \\\n",
       "0    XAACTCTAAATTTTCTCTTGGAAACTCCCATTTGAGATCATATTCA...  p.Y599_D600delinsX   \n",
       "1                                                TCTGC            p.L287fs   \n",
       "2                                            CGTACAAGA            p.R380fs   \n",
       "3                                                   CG            p.R370fs   \n",
       "4                                                    T             p.R882H   \n",
       "..                                                 ...                 ...   \n",
       "538                                                  G             p.K666N   \n",
       "539                                                  A             p.R882C   \n",
       "540                                                  A             p.R204X   \n",
       "541                                                CAA            p.H717fs   \n",
       "542                                                  A              p.S34F   \n",
       "\n",
       "    function       exonic function  mutated %  scDNA-seq VAF  \\\n",
       "0     exonic              stopgain   0.935200         0.7949   \n",
       "1     exonic  frameshift insertion   0.424900         0.4608   \n",
       "2     exonic  frameshift insertion   0.003400         0.0015   \n",
       "3     exonic  frameshift insertion   0.883600         0.4865   \n",
       "4     exonic     nonsynonymous SNV   0.760300         0.4639   \n",
       "..       ...                   ...        ...            ...   \n",
       "538   exonic     nonsynonymous SNV   0.903884         0.5040   \n",
       "539   exonic     nonsynonymous SNV   0.631245         0.3954   \n",
       "540   exonic              stopgain   0.420698         0.2909   \n",
       "541   exonic  frameshift insertion   0.706186         0.3574   \n",
       "542   exonic     nonsynonymous SNV   0.649881         0.3488   \n",
       "\n",
       "     WT\\n(No. of cells)  Het\\n(No. of cells)  Homo\\n(No. of cells)  \\\n",
       "0                   239                 8444                  2190   \n",
       "1                  1285                 4226                   606   \n",
       "2                 10779                   39                     0   \n",
       "3                   568                 9799                   248   \n",
       "4                   468                 5861                   169   \n",
       "..                  ...                  ...                   ...   \n",
       "538                 197                 3497                   133   \n",
       "539                 643                 1495                    97   \n",
       "540                1061                 1031                    30   \n",
       "541                 620                 1751                    30   \n",
       "542                 734                 1499                   140   \n",
       "\n",
       "     Missing\\n(No. of cells) validated          validation method  \\\n",
       "0                        498       yes                   bulk-NGS   \n",
       "1                       5254       yes                   bulk-NGS   \n",
       "2                        553       yes  bulk-NGS in paired sample   \n",
       "3                        756       yes                   bulk-NGS   \n",
       "4                       1433       yes                   bulk-NGS   \n",
       "..                       ...       ...                        ...   \n",
       "538                      189       yes        clinical sequencing   \n",
       "539                      287       yes        clinical sequencing   \n",
       "540                      400       yes        clinical sequencing   \n",
       "541                      121       yes        clinical sequencing   \n",
       "542                      149       yes        clinical sequencing   \n",
       "\n",
       "          bulk VAF  \n",
       "0           0.1186  \n",
       "1           0.1828  \n",
       "2                0  \n",
       "3            0.236  \n",
       "4           0.4239  \n",
       "..             ...  \n",
       "538  not available  \n",
       "539  not available  \n",
       "540  not available  \n",
       "541  not available  \n",
       "542  not available  \n",
       "\n",
       "[530 rows x 18 columns]"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "AML_variants"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Load haploid trinucleotide-context-site-specific mutation rates:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [],
   "source": [
    "site_mutation_rates = pd.read_csv(\"../data/AML_morita/Watson_2020_trinucleotide_mutation_rates.csv\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Site</th>\n",
       "      <th>Mutation_rate</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>A[C&gt;A]A</td>\n",
       "      <td>1.332230e-09</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>A[C&gt;A]C</td>\n",
       "      <td>9.481170e-10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>A[C&gt;A]G</td>\n",
       "      <td>1.064910e-09</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>A[C&gt;A]T</td>\n",
       "      <td>7.585960e-10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>C[C&gt;A]A</td>\n",
       "      <td>1.538010e-09</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>187</th>\n",
       "      <td>A[A&gt;C]C</td>\n",
       "      <td>1.688930e-10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>188</th>\n",
       "      <td>T[A&gt;C]A</td>\n",
       "      <td>1.554360e-10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>189</th>\n",
       "      <td>G[A&gt;C]A</td>\n",
       "      <td>1.574550e-10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>190</th>\n",
       "      <td>C[A&gt;C]A</td>\n",
       "      <td>2.836760e-10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>191</th>\n",
       "      <td>A[A&gt;C]A</td>\n",
       "      <td>1.857950e-10</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>192 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        Site  Mutation_rate\n",
       "0    A[C>A]A   1.332230e-09\n",
       "1    A[C>A]C   9.481170e-10\n",
       "2    A[C>A]G   1.064910e-09\n",
       "3    A[C>A]T   7.585960e-10\n",
       "4    C[C>A]A   1.538010e-09\n",
       "..       ...            ...\n",
       "187  A[A>C]C   1.688930e-10\n",
       "188  T[A>C]A   1.554360e-10\n",
       "189  G[A>C]A   1.574550e-10\n",
       "190  C[A>C]A   2.836760e-10\n",
       "191  A[A>C]A   1.857950e-10\n",
       "\n",
       "[192 rows x 2 columns]"
      ]
     },
     "execution_count": 35,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "site_mutation_rates"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Load human genome file (hg19):"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "hg19_fasta = Fasta(\"../data/AML_morita/GRCh37.p13.genome.fa\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "GCGG\n"
     ]
    }
   ],
   "source": [
    "chrom = 'chr2'\n",
    "start = 25457242 - 2\n",
    "end = 25457242 + 2\n",
    "\n",
    "# Extract the sequence\n",
    "sequence = hg19_fasta[chrom][start:end]\n",
    "print(sequence)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "For each variant, check whether it is an SNV. If yes, then query the trinucleotide context based on the columns \"chr\" and \"start\"."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ref = C, alt = T, queried = GCG, tri_context = G[C>T]G, tri_rate = 1.88229e-08\n",
      "ref = T, alt = A, queried = ATC, tri_context = A[T>A]C, tri_rate = 8.09824e-10\n",
      "ref = G, alt = T, queried = TGT, tri_context = T[G>T]T, tri_rate = 1.33223e-09\n",
      "ref = C, alt = G, queried = TCT, tri_context = T[C>G]T, tri_rate = 4.69475e-10\n",
      "ref = A, alt = T, queried = TAT, tri_context = T[A>T]T, tri_rate = 4.25782e-10\n",
      "ref = G, alt = A, queried = CGA, tri_context = C[G>A]A, tri_rate = 1.20044e-08\n",
      "ref = C, alt = T, queried = ACC, tri_context = A[C>T]C, tri_rate = 3.17573e-09\n",
      "ref = T, alt = A, queried = TTG, tri_context = T[T>A]G, tri_rate = 2.1756e-10\n",
      "ref = C, alt = G, queried = TCT, tri_context = T[C>G]T, tri_rate = 4.69475e-10\n",
      "ref = G, alt = T, queried = GGG, tri_context = G[G>T]G, tri_rate = 8.14186e-10\n",
      "ref = G, alt = A, queried = CGT, tri_context = C[G>A]T, tri_rate = 3.26929e-08\n",
      "ref = C, alt = T, queried = CCG, tri_context = C[C>T]G, tri_rate = 1.41512e-08\n",
      "ref = T, alt = G, queried = ATC, tri_context = A[T>G]C, tri_rate = 1.52343e-10\n",
      "ref = C, alt = A, queried = TCT, tri_context = T[C>A]T, tri_rate = 4.96095e-10\n",
      "ref = G, alt = T, queried = GGT, tri_context = G[G>T]T, tri_rate = 9.48117e-10\n",
      "ref = C, alt = T, queried = ACC, tri_context = A[C>T]C, tri_rate = 3.17573e-09\n",
      "ref = T, alt = G, queried = CTT, tri_context = C[T>G]T, tri_rate = 2.52155e-10\n",
      "ref = C, alt = T, queried = ACC, tri_context = A[C>T]C, tri_rate = 3.17573e-09\n",
      "ref = C, alt = G, queried = CCA, tri_context = C[C>G]A, tri_rate = 4.64303e-10\n",
      "ref = C, alt = T, queried = ACC, tri_context = A[C>T]C, tri_rate = 3.17573e-09\n",
      "ref = C, alt = T, queried = CCG, tri_context = C[C>T]G, tri_rate = 1.41512e-08\n",
      "ref = G, alt = T, queried = GGG, tri_context = G[G>T]G, tri_rate = 8.14186e-10\n",
      "ref = C, alt = T, queried = GCG, tri_context = G[C>T]G, tri_rate = 1.88229e-08\n",
      "ref = C, alt = T, queried = CCG, tri_context = C[C>T]G, tri_rate = 1.41512e-08\n",
      "ref = G, alt = A, queried = GGG, tri_context = G[G>A]G, tri_rate = 2.64611e-09\n",
      "ref = A, alt = T, queried = TAT, tri_context = T[A>T]T, tri_rate = 4.25782e-10\n",
      "ref = C, alt = A, queried = TCT, tri_context = T[C>A]T, tri_rate = 4.96095e-10\n",
      "ref = C, alt = T, queried = GCC, tri_context = G[C>T]C, tri_rate = 3.44567e-09\n",
      "ref = T, alt = C, queried = GTG, tri_context = G[T>C]G, tri_rate = 6.53787e-10\n",
      "ref = G, alt = A, queried = CGG, tri_context = C[G>A]G, tri_rate = 1.41512e-08\n",
      "ref = G, alt = A, queried = GGG, tri_context = G[G>A]G, tri_rate = 2.64611e-09\n",
      "ref = C, alt = G, queried = TCG, tri_context = T[C>G]G, tri_rate = 5.7233e-10\n",
      "ref = G, alt = A, queried = TGT, tri_context = T[G>A]T, tri_rate = 3.21114e-09\n",
      "ref = G, alt = C, queried = GGG, tri_context = G[G>C]G, tri_rate = 4.55944e-10\n",
      "ref = C, alt = T, queried = CCT, tri_context = C[C>T]T, tri_rate = 4.37909e-09\n",
      "ref = A, alt = T, queried = GAG, tri_context = G[A>T]G, tri_rate = 3.78353e-10\n",
      "ref = C, alt = T, queried = CCG, tri_context = C[C>T]G, tri_rate = 1.41512e-08\n",
      "ref = C, alt = G, queried = CCG, tri_context = C[C>G]G, tri_rate = 6.5522e-10\n",
      "ref = C, alt = T, queried = GCC, tri_context = G[C>T]C, tri_rate = 3.44567e-09\n",
      "ref = C, alt = A, queried = ACC, tri_context = A[C>A]C, tri_rate = 9.48117e-10\n",
      "ref = C, alt = T, queried = ACC, tri_context = A[C>T]C, tri_rate = 3.17573e-09\n",
      "ref = C, alt = G, queried = CCA, tri_context = C[C>G]A, tri_rate = 4.64303e-10\n",
      "ref = C, alt = T, queried = ACC, tri_context = A[C>T]C, tri_rate = 3.17573e-09\n",
      "ref = G, alt = A, queried = TGC, tri_context = T[G>A]C, tri_rate = 2.32279e-09\n",
      "ref = G, alt = C, queried = CGA, tri_context = C[G>C]A, tri_rate = 5.7233e-10\n",
      "ref = C, alt = T, queried = GCG, tri_context = G[C>T]G, tri_rate = 1.88229e-08\n",
      "ref = C, alt = T, queried = ACG, tri_context = A[C>T]G, tri_rate = 3.26929e-08\n",
      "ref = C, alt = T, queried = CCG, tri_context = C[C>T]G, tri_rate = 1.41512e-08\n",
      "ref = A, alt = T, queried = CAA, tri_context = C[A>T]A, tri_rate = 2.1756e-10\n",
      "ref = C, alt = T, queried = ACC, tri_context = A[C>T]C, tri_rate = 3.17573e-09\n",
      "ref = C, alt = T, queried = CCT, tri_context = C[C>T]T, tri_rate = 4.37909e-09\n",
      "ref = G, alt = A, queried = CGG, tri_context = C[G>A]G, tri_rate = 1.41512e-08\n",
      "ref = C, alt = A, queried = ACC, tri_context = A[C>A]C, tri_rate = 9.48117e-10\n",
      "ref = C, alt = T, queried = ACC, tri_context = A[C>T]C, tri_rate = 3.17573e-09\n",
      "ref = C, alt = A, queried = ACC, tri_context = A[C>A]C, tri_rate = 9.48117e-10\n",
      "ref = C, alt = T, queried = CCA, tri_context = C[C>T]A, tri_rate = 1.96458e-09\n",
      "ref = C, alt = T, queried = ACC, tri_context = A[C>T]C, tri_rate = 3.17573e-09\n",
      "ref = C, alt = A, queried = CCA, tri_context = C[C>A]A, tri_rate = 1.53801e-09\n",
      "ref = C, alt = A, queried = ACC, tri_context = A[C>A]C, tri_rate = 9.48117e-10\n",
      "ref = C, alt = T, queried = ACC, tri_context = A[C>T]C, tri_rate = 3.17573e-09\n",
      "ref = C, alt = A, queried = CCT, tri_context = C[C>A]T, tri_rate = 9.82508e-10\n",
      "ref = C, alt = T, queried = CCT, tri_context = C[C>T]T, tri_rate = 4.37909e-09\n",
      "ref = C, alt = G, queried = CCA, tri_context = C[C>G]A, tri_rate = 4.64303e-10\n",
      "ref = C, alt = T, queried = GCG, tri_context = G[C>T]G, tri_rate = 1.88229e-08\n",
      "ref = G, alt = A, queried = CGC, tri_context = C[G>A]C, tri_rate = 1.88229e-08\n",
      "ref = G, alt = T, queried = AGA, tri_context = A[G>T]A, tri_rate = 4.96095e-10\n",
      "ref = C, alt = T, queried = ACA, tri_context = A[C>T]A, tri_rate = 3.21114e-09\n",
      "ref = G, alt = A, queried = CGG, tri_context = C[G>A]G, tri_rate = 1.41512e-08\n",
      "ref = C, alt = T, queried = GCG, tri_context = G[C>T]G, tri_rate = 1.88229e-08\n",
      "ref = T, alt = G, queried = ATC, tri_context = A[T>G]C, tri_rate = 1.52343e-10\n",
      "ref = C, alt = A, queried = TCT, tri_context = T[C>A]T, tri_rate = 4.96095e-10\n",
      "ref = T, alt = A, queried = TTG, tri_context = T[T>A]G, tri_rate = 2.1756e-10\n",
      "ref = C, alt = T, queried = ACC, tri_context = A[C>T]C, tri_rate = 3.17573e-09\n",
      "ref = G, alt = A, queried = CGG, tri_context = C[G>A]G, tri_rate = 1.41512e-08\n",
      "ref = G, alt = A, queried = CGG, tri_context = C[G>A]G, tri_rate = 1.41512e-08\n",
      "ref = C, alt = T, queried = ACG, tri_context = A[C>T]G, tri_rate = 3.26929e-08\n",
      "ref = C, alt = T, queried = CCG, tri_context = C[C>T]G, tri_rate = 1.41512e-08\n",
      "ref = C, alt = T, queried = CCG, tri_context = C[C>T]G, tri_rate = 1.41512e-08\n",
      "ref = C, alt = T, queried = TCG, tri_context = T[C>T]G, tri_rate = 1.20044e-08\n",
      "ref = G, alt = T, queried = GGG, tri_context = G[G>T]G, tri_rate = 8.14186e-10\n",
      "ref = C, alt = T, queried = GCG, tri_context = G[C>T]G, tri_rate = 1.88229e-08\n",
      "ref = C, alt = T, queried = CCG, tri_context = C[C>T]G, tri_rate = 1.41512e-08\n",
      "ref = C, alt = T, queried = ACC, tri_context = A[C>T]C, tri_rate = 3.17573e-09\n",
      "ref = C, alt = A, queried = ACC, tri_context = A[C>A]C, tri_rate = 9.48117e-10\n",
      "ref = C, alt = A, queried = ACC, tri_context = A[C>A]C, tri_rate = 9.48117e-10\n",
      "ref = C, alt = T, queried = GCG, tri_context = G[C>T]G, tri_rate = 1.88229e-08\n",
      "ref = C, alt = A, queried = TCT, tri_context = T[C>A]T, tri_rate = 4.96095e-10\n",
      "ref = G, alt = C, queried = CGA, tri_context = C[G>C]A, tri_rate = 5.7233e-10\n",
      "ref = T, alt = C, queried = TTC, tri_context = T[T>C]C, tri_rate = 5.40234e-10\n",
      "ref = C, alt = T, queried = CCG, tri_context = C[C>T]G, tri_rate = 1.41512e-08\n",
      "ref = C, alt = T, queried = TCC, tri_context = T[C>T]C, tri_rate = 1.56027e-09\n",
      "ref = G, alt = T, queried = GGG, tri_context = G[G>T]G, tri_rate = 8.14186e-10\n",
      "ref = C, alt = T, queried = ACC, tri_context = A[C>T]C, tri_rate = 3.17573e-09\n",
      "ref = C, alt = T, queried = ACC, tri_context = A[C>T]C, tri_rate = 3.17573e-09\n",
      "ref = C, alt = T, queried = ACC, tri_context = A[C>T]C, tri_rate = 3.17573e-09\n",
      "ref = T, alt = C, queried = TTC, tri_context = T[T>C]C, tri_rate = 5.40234e-10\n",
      "ref = C, alt = T, queried = ACC, tri_context = A[C>T]C, tri_rate = 3.17573e-09\n",
      "ref = C, alt = T, queried = GCC, tri_context = G[C>T]C, tri_rate = 3.44567e-09\n",
      "ref = A, alt = T, queried = TAT, tri_context = T[A>T]T, tri_rate = 4.25782e-10\n",
      "ref = C, alt = T, queried = ACC, tri_context = A[C>T]C, tri_rate = 3.17573e-09\n",
      "ref = C, alt = T, queried = CCT, tri_context = C[C>T]T, tri_rate = 4.37909e-09\n",
      "ref = T, alt = C, queried = TTG, tri_context = T[T>C]G, tri_rate = 5.44182e-10\n",
      "ref = C, alt = T, queried = ACC, tri_context = A[C>T]C, tri_rate = 3.17573e-09\n",
      "ref = C, alt = G, queried = TCT, tri_context = T[C>G]T, tri_rate = 4.69475e-10\n",
      "ref = C, alt = T, queried = ACG, tri_context = A[C>T]G, tri_rate = 3.26929e-08\n",
      "ref = C, alt = T, queried = CCG, tri_context = C[C>T]G, tri_rate = 1.41512e-08\n",
      "ref = C, alt = G, queried = ACC, tri_context = A[C>G]C, tri_rate = 4.83263e-10\n",
      "ref = C, alt = T, queried = ACC, tri_context = A[C>T]C, tri_rate = 3.17573e-09\n",
      "ref = C, alt = G, queried = CCA, tri_context = C[C>G]A, tri_rate = 4.64303e-10\n",
      "ref = C, alt = G, queried = ACC, tri_context = A[C>G]C, tri_rate = 4.83263e-10\n",
      "ref = G, alt = C, queried = TGA, tri_context = T[G>C]A, tri_rate = 2.83107e-10\n",
      "ref = C, alt = G, queried = GCC, tri_context = G[C>G]C, tri_rate = 4.04844e-10\n",
      "ref = G, alt = C, queried = GGG, tri_context = G[G>C]G, tri_rate = 4.55944e-10\n",
      "ref = T, alt = C, queried = GTT, tri_context = G[T>C]T, tri_rate = 1.09413e-09\n",
      "ref = C, alt = T, queried = GCC, tri_context = G[C>T]C, tri_rate = 3.44567e-09\n",
      "ref = G, alt = C, queried = GGG, tri_context = G[G>C]G, tri_rate = 4.55944e-10\n",
      "ref = T, alt = C, queried = TTC, tri_context = T[T>C]C, tri_rate = 5.40234e-10\n",
      "ref = G, alt = C, queried = GGT, tri_context = G[G>C]T, tri_rate = 4.83263e-10\n",
      "ref = G, alt = T, queried = GGT, tri_context = G[G>T]T, tri_rate = 9.48117e-10\n",
      "ref = C, alt = T, queried = CCG, tri_context = C[C>T]G, tri_rate = 1.41512e-08\n",
      "ref = C, alt = T, queried = ACC, tri_context = A[C>T]C, tri_rate = 3.17573e-09\n",
      "ref = C, alt = T, queried = ACC, tri_context = A[C>T]C, tri_rate = 3.17573e-09\n",
      "ref = C, alt = T, queried = CCG, tri_context = C[C>T]G, tri_rate = 1.41512e-08\n",
      "ref = C, alt = G, queried = ACC, tri_context = A[C>G]C, tri_rate = 4.83263e-10\n",
      "ref = C, alt = G, queried = CCA, tri_context = C[C>G]A, tri_rate = 4.64303e-10\n",
      "ref = C, alt = G, queried = ACC, tri_context = A[C>G]C, tri_rate = 4.83263e-10\n",
      "ref = G, alt = C, queried = TGA, tri_context = T[G>C]A, tri_rate = 2.83107e-10\n",
      "ref = G, alt = A, queried = TGC, tri_context = T[G>A]C, tri_rate = 2.32279e-09\n",
      "ref = G, alt = T, queried = GGG, tri_context = G[G>T]G, tri_rate = 8.14186e-10\n",
      "ref = T, alt = C, queried = GTT, tri_context = G[T>C]T, tri_rate = 1.09413e-09\n",
      "ref = C, alt = T, queried = ACG, tri_context = A[C>T]G, tri_rate = 3.26929e-08\n",
      "ref = C, alt = A, queried = ACC, tri_context = A[C>A]C, tri_rate = 9.48117e-10\n",
      "ref = C, alt = T, queried = ACC, tri_context = A[C>T]C, tri_rate = 3.17573e-09\n",
      "ref = C, alt = T, queried = ACC, tri_context = A[C>T]C, tri_rate = 3.17573e-09\n",
      "ref = C, alt = A, queried = CCG, tri_context = C[C>A]G, tri_rate = 1.16311e-09\n",
      "ref = C, alt = G, queried = CCA, tri_context = C[C>G]A, tri_rate = 4.64303e-10\n",
      "ref = C, alt = G, queried = ACC, tri_context = A[C>G]C, tri_rate = 4.83263e-10\n",
      "ref = C, alt = T, queried = ACC, tri_context = A[C>T]C, tri_rate = 3.17573e-09\n",
      "ref = A, alt = G, queried = CAC, tri_context = C[A>G]C, tri_rate = 6.53787e-10\n",
      "ref = G, alt = A, queried = TGC, tri_context = T[G>A]C, tri_rate = 2.32279e-09\n",
      "ref = C, alt = A, queried = TCT, tri_context = T[C>A]T, tri_rate = 4.96095e-10\n",
      "ref = C, alt = T, queried = CCT, tri_context = C[C>T]T, tri_rate = 4.37909e-09\n",
      "ref = C, alt = T, queried = ACG, tri_context = A[C>T]G, tri_rate = 3.26929e-08\n",
      "ref = C, alt = T, queried = ACC, tri_context = A[C>T]C, tri_rate = 3.17573e-09\n",
      "ref = T, alt = C, queried = ATA, tri_context = A[T>C]A, tri_rate = 1.47206e-09\n",
      "ref = G, alt = A, queried = CGG, tri_context = C[G>A]G, tri_rate = 1.41512e-08\n",
      "ref = C, alt = A, queried = TCT, tri_context = T[C>A]T, tri_rate = 4.96095e-10\n",
      "ref = C, alt = A, queried = ACC, tri_context = A[C>A]C, tri_rate = 9.48117e-10\n",
      "ref = C, alt = T, queried = ACC, tri_context = A[C>T]C, tri_rate = 3.17573e-09\n",
      "ref = C, alt = T, queried = CCG, tri_context = C[C>T]G, tri_rate = 1.41512e-08\n",
      "ref = C, alt = G, queried = CCA, tri_context = C[C>G]A, tri_rate = 4.64303e-10\n",
      "ref = C, alt = T, queried = ACC, tri_context = A[C>T]C, tri_rate = 3.17573e-09\n",
      "ref = G, alt = C, queried = GGG, tri_context = G[G>C]G, tri_rate = 4.55944e-10\n",
      "ref = G, alt = A, queried = CGG, tri_context = C[G>A]G, tri_rate = 1.41512e-08\n",
      "ref = G, alt = A, queried = CGA, tri_context = C[G>A]A, tri_rate = 1.20044e-08\n",
      "ref = C, alt = T, queried = TCG, tri_context = T[C>T]G, tri_rate = 1.20044e-08\n",
      "ref = G, alt = T, queried = GGG, tri_context = G[G>T]G, tri_rate = 8.14186e-10\n",
      "ref = C, alt = T, queried = CCG, tri_context = C[C>T]G, tri_rate = 1.41512e-08\n",
      "ref = G, alt = C, queried = GGG, tri_context = G[G>C]G, tri_rate = 4.55944e-10\n",
      "ref = C, alt = T, queried = GCG, tri_context = G[C>T]G, tri_rate = 1.88229e-08\n",
      "ref = C, alt = T, queried = CCG, tri_context = C[C>T]G, tri_rate = 1.41512e-08\n",
      "ref = G, alt = T, queried = GGG, tri_context = G[G>T]G, tri_rate = 8.14186e-10\n",
      "ref = C, alt = T, queried = GCG, tri_context = G[C>T]G, tri_rate = 1.88229e-08\n",
      "ref = G, alt = A, queried = CGA, tri_context = C[G>A]A, tri_rate = 1.20044e-08\n",
      "ref = C, alt = T, queried = TCC, tri_context = T[C>T]C, tri_rate = 1.56027e-09\n",
      "ref = C, alt = T, queried = CCG, tri_context = C[C>T]G, tri_rate = 1.41512e-08\n",
      "ref = G, alt = T, queried = GGG, tri_context = G[G>T]G, tri_rate = 8.14186e-10\n",
      "ref = C, alt = T, queried = ACG, tri_context = A[C>T]G, tri_rate = 3.26929e-08\n",
      "ref = C, alt = T, queried = CCG, tri_context = C[C>T]G, tri_rate = 1.41512e-08\n",
      "ref = C, alt = T, queried = ACC, tri_context = A[C>T]C, tri_rate = 3.17573e-09\n",
      "ref = A, alt = G, queried = GAA, tri_context = G[A>G]A, tri_rate = 5.40234e-10\n",
      "ref = C, alt = T, queried = TCC, tri_context = T[C>T]C, tri_rate = 1.56027e-09\n",
      "ref = C, alt = T, queried = ACG, tri_context = A[C>T]G, tri_rate = 3.26929e-08\n",
      "ref = A, alt = C, queried = GAT, tri_context = G[A>C]T, tri_rate = 1.52343e-10\n",
      "ref = T, alt = C, queried = ATT, tri_context = A[T>C]T, tri_rate = 1.34625e-09\n",
      "ref = G, alt = A, queried = TGC, tri_context = T[G>A]C, tri_rate = 2.32279e-09\n",
      "ref = C, alt = T, queried = CCG, tri_context = C[C>T]G, tri_rate = 1.41512e-08\n",
      "ref = A, alt = T, queried = GAC, tri_context = G[A>T]C, tri_rate = 4.53317e-10\n",
      "ref = C, alt = T, queried = GCG, tri_context = G[C>T]G, tri_rate = 1.88229e-08\n",
      "ref = C, alt = T, queried = CCG, tri_context = C[C>T]G, tri_rate = 1.41512e-08\n",
      "ref = G, alt = T, queried = TGT, tri_context = T[G>T]T, tri_rate = 1.33223e-09\n",
      "ref = G, alt = T, queried = GGG, tri_context = G[G>T]G, tri_rate = 8.14186e-10\n",
      "ref = C, alt = T, queried = GCG, tri_context = G[C>T]G, tri_rate = 1.88229e-08\n",
      "ref = C, alt = T, queried = CCT, tri_context = C[C>T]T, tri_rate = 4.37909e-09\n",
      "ref = G, alt = A, queried = GGG, tri_context = G[G>A]G, tri_rate = 2.64611e-09\n",
      "ref = G, alt = A, queried = GGG, tri_context = G[G>A]G, tri_rate = 2.64611e-09\n",
      "ref = C, alt = T, queried = CCG, tri_context = C[C>T]G, tri_rate = 1.41512e-08\n",
      "ref = T, alt = A, queried = CTT, tri_context = C[T>A]T, tri_rate = 3.83597e-10\n",
      "ref = C, alt = A, queried = ACC, tri_context = A[C>A]C, tri_rate = 9.48117e-10\n",
      "ref = C, alt = T, queried = CCT, tri_context = C[C>T]T, tri_rate = 4.37909e-09\n",
      "ref = G, alt = A, queried = CGC, tri_context = C[G>A]C, tri_rate = 1.88229e-08\n",
      "ref = T, alt = C, queried = CTG, tri_context = C[T>C]G, tri_rate = 8.05685e-10\n",
      "ref = G, alt = A, queried = AGA, tri_context = A[G>A]A, tri_rate = 1.39633e-09\n",
      "ref = C, alt = T, queried = GCG, tri_context = G[C>T]G, tri_rate = 1.88229e-08\n",
      "ref = C, alt = T, queried = CCA, tri_context = C[C>T]A, tri_rate = 1.96458e-09\n",
      "ref = C, alt = G, queried = ACC, tri_context = A[C>G]C, tri_rate = 4.83263e-10\n",
      "ref = C, alt = T, queried = ACC, tri_context = A[C>T]C, tri_rate = 3.17573e-09\n",
      "ref = C, alt = T, queried = CCG, tri_context = C[C>T]G, tri_rate = 1.41512e-08\n",
      "ref = A, alt = T, queried = GAT, tri_context = G[A>T]T, tri_rate = 8.09824e-10\n",
      "ref = C, alt = T, queried = TCC, tri_context = T[C>T]C, tri_rate = 1.56027e-09\n",
      "ref = G, alt = A, queried = GGG, tri_context = G[G>A]G, tri_rate = 2.64611e-09\n",
      "ref = C, alt = T, queried = CCG, tri_context = C[C>T]G, tri_rate = 1.41512e-08\n",
      "ref = C, alt = G, queried = CCT, tri_context = C[C>G]T, tri_rate = 4.37005e-10\n",
      "ref = C, alt = T, queried = TCA, tri_context = T[C>T]A, tri_rate = 9.91969e-10\n",
      "ref = G, alt = C, queried = TGA, tri_context = T[G>C]A, tri_rate = 2.83107e-10\n",
      "ref = G, alt = A, queried = GGG, tri_context = G[G>A]G, tri_rate = 2.64611e-09\n",
      "ref = C, alt = A, queried = ACT, tri_context = A[C>A]T, tri_rate = 7.58596e-10\n",
      "ref = C, alt = G, queried = TCT, tri_context = T[C>G]T, tri_rate = 4.69475e-10\n",
      "ref = C, alt = T, queried = ACC, tri_context = A[C>T]C, tri_rate = 3.17573e-09\n",
      "ref = C, alt = T, queried = ACC, tri_context = A[C>T]C, tri_rate = 3.17573e-09\n",
      "ref = A, alt = C, queried = GAT, tri_context = G[A>C]T, tri_rate = 1.52343e-10\n",
      "ref = G, alt = A, queried = TGA, tri_context = T[G>A]A, tri_rate = 9.91969e-10\n",
      "ref = C, alt = G, queried = CCT, tri_context = C[C>G]T, tri_rate = 4.37005e-10\n",
      "ref = G, alt = T, queried = GGG, tri_context = G[G>T]G, tri_rate = 8.14186e-10\n",
      "ref = C, alt = G, queried = GCG, tri_context = G[C>G]G, tri_rate = 6.29677e-10\n",
      "ref = T, alt = C, queried = ATC, tri_context = A[T>C]C, tri_rate = 8.25861e-10\n",
      "ref = T, alt = A, queried = CTT, tri_context = C[T>A]T, tri_rate = 3.83597e-10\n",
      "ref = C, alt = G, queried = CCA, tri_context = C[C>G]A, tri_rate = 4.64303e-10\n",
      "ref = C, alt = G, queried = ACC, tri_context = A[C>G]C, tri_rate = 4.83263e-10\n",
      "ref = G, alt = A, queried = AGA, tri_context = A[G>A]A, tri_rate = 1.39633e-09\n",
      "ref = G, alt = A, queried = CGA, tri_context = C[G>A]A, tri_rate = 1.20044e-08\n",
      "ref = C, alt = T, queried = GCG, tri_context = G[C>T]G, tri_rate = 1.88229e-08\n",
      "ref = C, alt = T, queried = CCT, tri_context = C[C>T]T, tri_rate = 4.37909e-09\n",
      "ref = C, alt = A, queried = ACC, tri_context = A[C>A]C, tri_rate = 9.48117e-10\n",
      "ref = G, alt = A, queried = CGA, tri_context = C[G>A]A, tri_rate = 1.20044e-08\n",
      "ref = T, alt = C, queried = CTG, tri_context = C[T>C]G, tri_rate = 8.05685e-10\n",
      "ref = C, alt = T, queried = GCG, tri_context = G[C>T]G, tri_rate = 1.88229e-08\n",
      "ref = C, alt = T, queried = ACA, tri_context = A[C>T]A, tri_rate = 3.21114e-09\n",
      "ref = A, alt = T, queried = GAG, tri_context = G[A>T]G, tri_rate = 3.78353e-10\n",
      "ref = G, alt = A, queried = GGC, tri_context = G[G>A]C, tri_rate = 3.44567e-09\n",
      "ref = C, alt = A, queried = GCC, tri_context = G[C>A]C, tri_rate = 6.41452e-10\n",
      "ref = A, alt = C, queried = TAT, tri_context = T[A>C]T, tri_rate = 1.9991e-10\n",
      "ref = C, alt = A, queried = TCT, tri_context = T[C>A]T, tri_rate = 4.96095e-10\n",
      "ref = C, alt = T, queried = GCC, tri_context = G[C>T]C, tri_rate = 3.44567e-09\n",
      "ref = C, alt = T, queried = ACC, tri_context = A[C>T]C, tri_rate = 3.17573e-09\n",
      "ref = C, alt = A, queried = CCT, tri_context = C[C>A]T, tri_rate = 9.82508e-10\n",
      "ref = C, alt = T, queried = CCT, tri_context = C[C>T]T, tri_rate = 4.37909e-09\n",
      "ref = G, alt = A, queried = CGG, tri_context = C[G>A]G, tri_rate = 1.41512e-08\n",
      "ref = C, alt = A, queried = TCT, tri_context = T[C>A]T, tri_rate = 4.96095e-10\n",
      "ref = C, alt = T, queried = ACG, tri_context = A[C>T]G, tri_rate = 3.26929e-08\n",
      "ref = G, alt = T, queried = GGT, tri_context = G[G>T]T, tri_rate = 9.48117e-10\n",
      "ref = A, alt = T, queried = GAG, tri_context = G[A>T]G, tri_rate = 3.78353e-10\n",
      "ref = A, alt = G, queried = CAG, tri_context = C[A>G]G, tri_rate = 8.05685e-10\n",
      "ref = C, alt = T, queried = CCT, tri_context = C[C>T]T, tri_rate = 4.37909e-09\n",
      "ref = C, alt = T, queried = CCG, tri_context = C[C>T]G, tri_rate = 1.41512e-08\n",
      "ref = C, alt = T, queried = GCG, tri_context = G[C>T]G, tri_rate = 1.88229e-08\n",
      "ref = G, alt = A, queried = AGG, tri_context = A[G>A]G, tri_rate = 4.37909e-09\n",
      "ref = C, alt = A, queried = CCA, tri_context = C[C>A]A, tri_rate = 1.53801e-09\n",
      "ref = C, alt = G, queried = CCC, tri_context = C[C>G]C, tri_rate = 4.55944e-10\n",
      "ref = C, alt = T, queried = CCG, tri_context = C[C>T]G, tri_rate = 1.41512e-08\n",
      "ref = C, alt = A, queried = CCG, tri_context = C[C>A]G, tri_rate = 1.16311e-09\n",
      "ref = G, alt = A, queried = CGG, tri_context = C[G>A]G, tri_rate = 1.41512e-08\n",
      "ref = C, alt = T, queried = GCG, tri_context = G[C>T]G, tri_rate = 1.88229e-08\n",
      "ref = T, alt = A, queried = ATC, tri_context = A[T>A]C, tri_rate = 8.09824e-10\n",
      "ref = T, alt = C, queried = CTT, tri_context = C[T>C]T, tri_rate = 7.36078e-10\n",
      "ref = A, alt = C, queried = AAG, tri_context = A[A>C]G, tri_rate = 2.52155e-10\n",
      "ref = G, alt = T, queried = TGT, tri_context = T[G>T]T, tri_rate = 1.33223e-09\n",
      "ref = T, alt = A, queried = CTT, tri_context = C[T>A]T, tri_rate = 3.83597e-10\n",
      "ref = G, alt = C, queried = TGT, tri_context = T[G>C]T, tri_rate = 1.08807e-09\n",
      "ref = C, alt = T, queried = ACC, tri_context = A[C>T]C, tri_rate = 3.17573e-09\n",
      "ref = G, alt = A, queried = CGA, tri_context = C[G>A]A, tri_rate = 1.20044e-08\n",
      "ref = C, alt = A, queried = TCT, tri_context = T[C>A]T, tri_rate = 4.96095e-10\n",
      "ref = C, alt = G, queried = ACA, tri_context = A[C>G]A, tri_rate = 1.08807e-09\n",
      "ref = A, alt = T, queried = GAC, tri_context = G[A>T]C, tri_rate = 4.53317e-10\n",
      "ref = C, alt = T, queried = GCC, tri_context = G[C>T]C, tri_rate = 3.44567e-09\n",
      "ref = G, alt = C, queried = GGG, tri_context = G[G>C]G, tri_rate = 4.55944e-10\n",
      "ref = C, alt = T, queried = ACG, tri_context = A[C>T]G, tri_rate = 3.26929e-08\n",
      "ref = G, alt = A, queried = CGA, tri_context = C[G>A]A, tri_rate = 1.20044e-08\n",
      "ref = G, alt = T, queried = TGT, tri_context = T[G>T]T, tri_rate = 1.33223e-09\n",
      "ref = C, alt = T, queried = TCG, tri_context = T[C>T]G, tri_rate = 1.20044e-08\n",
      "ref = C, alt = T, queried = CCG, tri_context = C[C>T]G, tri_rate = 1.41512e-08\n",
      "ref = G, alt = A, queried = GGG, tri_context = G[G>A]G, tri_rate = 2.64611e-09\n",
      "ref = G, alt = A, queried = CGA, tri_context = C[G>A]A, tri_rate = 1.20044e-08\n",
      "ref = G, alt = C, queried = GGG, tri_context = G[G>C]G, tri_rate = 4.55944e-10\n",
      "ref = G, alt = C, queried = GGG, tri_context = G[G>C]G, tri_rate = 4.55944e-10\n",
      "ref = C, alt = T, queried = ACG, tri_context = A[C>T]G, tri_rate = 3.26929e-08\n",
      "ref = G, alt = A, queried = CGA, tri_context = C[G>A]A, tri_rate = 1.20044e-08\n",
      "ref = A, alt = G, queried = AAA, tri_context = A[A>G]A, tri_rate = 4.95824e-10\n",
      "ref = A, alt = G, queried = AAC, tri_context = A[A>G]C, tri_rate = 1.09413e-09\n",
      "ref = C, alt = A, queried = ACC, tri_context = A[C>A]C, tri_rate = 9.48117e-10\n",
      "ref = C, alt = T, queried = ACC, tri_context = A[C>T]C, tri_rate = 3.17573e-09\n",
      "ref = C, alt = G, queried = CCT, tri_context = C[C>G]T, tri_rate = 4.37005e-10\n",
      "ref = C, alt = T, queried = CCT, tri_context = C[C>T]T, tri_rate = 4.37909e-09\n",
      "ref = C, alt = T, queried = GCG, tri_context = G[C>T]G, tri_rate = 1.88229e-08\n",
      "ref = C, alt = T, queried = ACC, tri_context = A[C>T]C, tri_rate = 3.17573e-09\n",
      "ref = A, alt = T, queried = GAC, tri_context = G[A>T]C, tri_rate = 4.53317e-10\n",
      "ref = G, alt = A, queried = TGC, tri_context = T[G>A]C, tri_rate = 2.32279e-09\n",
      "ref = C, alt = T, queried = GCG, tri_context = G[C>T]G, tri_rate = 1.88229e-08\n",
      "ref = C, alt = T, queried = CCG, tri_context = C[C>T]G, tri_rate = 1.41512e-08\n",
      "ref = C, alt = T, queried = TCC, tri_context = T[C>T]C, tri_rate = 1.56027e-09\n",
      "ref = C, alt = T, queried = CCT, tri_context = C[C>T]T, tri_rate = 4.37909e-09\n",
      "ref = G, alt = A, queried = GGG, tri_context = G[G>A]G, tri_rate = 2.64611e-09\n",
      "ref = C, alt = T, queried = GCG, tri_context = G[C>T]G, tri_rate = 1.88229e-08\n",
      "ref = C, alt = T, queried = CCG, tri_context = C[C>T]G, tri_rate = 1.41512e-08\n",
      "ref = C, alt = T, queried = ACC, tri_context = A[C>T]C, tri_rate = 3.17573e-09\n",
      "ref = A, alt = G, queried = CAG, tri_context = C[A>G]G, tri_rate = 8.05685e-10\n",
      "ref = G, alt = C, queried = GGT, tri_context = G[G>C]T, tri_rate = 4.83263e-10\n",
      "ref = A, alt = G, queried = AAA, tri_context = A[A>G]A, tri_rate = 4.95824e-10\n",
      "ref = C, alt = T, queried = GCC, tri_context = G[C>T]C, tri_rate = 3.44567e-09\n",
      "ref = T, alt = C, queried = ATA, tri_context = A[T>C]A, tri_rate = 1.47206e-09\n",
      "ref = C, alt = T, queried = ACA, tri_context = A[C>T]A, tri_rate = 3.21114e-09\n",
      "ref = C, alt = T, queried = GCG, tri_context = G[C>T]G, tri_rate = 1.88229e-08\n",
      "ref = C, alt = A, queried = TCT, tri_context = T[C>A]T, tri_rate = 4.96095e-10\n",
      "ref = G, alt = A, queried = CGG, tri_context = C[G>A]G, tri_rate = 1.41512e-08\n",
      "ref = G, alt = T, queried = TGT, tri_context = T[G>T]T, tri_rate = 1.33223e-09\n",
      "ref = G, alt = A, queried = AGA, tri_context = A[G>A]A, tri_rate = 1.39633e-09\n",
      "ref = C, alt = T, queried = CCG, tri_context = C[C>T]G, tri_rate = 1.41512e-08\n",
      "ref = G, alt = A, queried = CGG, tri_context = C[G>A]G, tri_rate = 1.41512e-08\n",
      "ref = C, alt = T, queried = CCT, tri_context = C[C>T]T, tri_rate = 4.37909e-09\n",
      "ref = G, alt = A, queried = AGA, tri_context = A[G>A]A, tri_rate = 1.39633e-09\n",
      "ref = C, alt = T, queried = TCT, tri_context = T[C>T]T, tri_rate = 1.39633e-09\n",
      "ref = G, alt = A, queried = CGG, tri_context = C[G>A]G, tri_rate = 1.41512e-08\n",
      "ref = C, alt = T, queried = CCG, tri_context = C[C>T]G, tri_rate = 1.41512e-08\n",
      "ref = G, alt = T, queried = GGG, tri_context = G[G>T]G, tri_rate = 8.14186e-10\n",
      "ref = C, alt = T, queried = CCG, tri_context = C[C>T]G, tri_rate = 1.41512e-08\n",
      "ref = G, alt = T, queried = GGG, tri_context = G[G>T]G, tri_rate = 8.14186e-10\n",
      "ref = C, alt = A, queried = ACC, tri_context = A[C>A]C, tri_rate = 9.48117e-10\n",
      "ref = C, alt = T, queried = GCA, tri_context = G[C>T]A, tri_rate = 2.32279e-09\n",
      "ref = C, alt = G, queried = CCG, tri_context = C[C>G]G, tri_rate = 6.5522e-10\n",
      "ref = G, alt = A, queried = CGA, tri_context = C[G>A]A, tri_rate = 1.20044e-08\n",
      "ref = G, alt = A, queried = CGA, tri_context = C[G>A]A, tri_rate = 1.20044e-08\n",
      "ref = C, alt = T, queried = GCG, tri_context = G[C>T]G, tri_rate = 1.88229e-08\n",
      "ref = A, alt = T, queried = GAC, tri_context = G[A>T]C, tri_rate = 4.53317e-10\n",
      "ref = C, alt = T, queried = CCC, tri_context = C[C>T]C, tri_rate = 2.64611e-09\n",
      "ref = C, alt = T, queried = ACG, tri_context = A[C>T]G, tri_rate = 3.26929e-08\n",
      "ref = G, alt = A, queried = CGG, tri_context = C[G>A]G, tri_rate = 1.41512e-08\n",
      "ref = C, alt = A, queried = CCT, tri_context = C[C>A]T, tri_rate = 9.82508e-10\n",
      "ref = G, alt = A, queried = CGA, tri_context = C[G>A]A, tri_rate = 1.20044e-08\n",
      "ref = C, alt = T, queried = GCG, tri_context = G[C>T]G, tri_rate = 1.88229e-08\n",
      "ref = C, alt = T, queried = CCG, tri_context = C[C>T]G, tri_rate = 1.41512e-08\n",
      "ref = G, alt = C, queried = GGG, tri_context = G[G>C]G, tri_rate = 4.55944e-10\n",
      "ref = C, alt = T, queried = GCG, tri_context = G[C>T]G, tri_rate = 1.88229e-08\n",
      "ref = C, alt = T, queried = CCG, tri_context = C[C>T]G, tri_rate = 1.41512e-08\n",
      "ref = C, alt = T, queried = ACA, tri_context = A[C>T]A, tri_rate = 3.21114e-09\n",
      "ref = G, alt = C, queried = GGG, tri_context = G[G>C]G, tri_rate = 4.55944e-10\n",
      "ref = G, alt = A, queried = GGT, tri_context = G[G>A]T, tri_rate = 3.17573e-09\n",
      "ref = C, alt = T, queried = ACG, tri_context = A[C>T]G, tri_rate = 3.26929e-08\n",
      "ref = C, alt = G, queried = TCT, tri_context = T[C>G]T, tri_rate = 4.69475e-10\n",
      "ref = A, alt = G, queried = GAA, tri_context = G[A>G]A, tri_rate = 5.40234e-10\n",
      "ref = C, alt = A, queried = TCT, tri_context = T[C>A]T, tri_rate = 4.96095e-10\n",
      "ref = C, alt = T, queried = GCG, tri_context = G[C>T]G, tri_rate = 1.88229e-08\n",
      "ref = C, alt = G, queried = TCT, tri_context = T[C>G]T, tri_rate = 4.69475e-10\n",
      "ref = T, alt = G, queried = GTT, tri_context = G[T>G]T, tri_rate = 1.68893e-10\n",
      "ref = C, alt = T, queried = ACC, tri_context = A[C>T]C, tri_rate = 3.17573e-09\n",
      "ref = C, alt = G, queried = TCT, tri_context = T[C>G]T, tri_rate = 4.69475e-10\n",
      "ref = G, alt = A, queried = CGG, tri_context = C[G>A]G, tri_rate = 1.41512e-08\n",
      "ref = G, alt = A, queried = AGA, tri_context = A[G>A]A, tri_rate = 1.39633e-09\n"
     ]
    }
   ],
   "source": [
    "# For each variant, check whether it is an SNV. If yes, then query the trinucleotide context based on the columns \"chr\" and \"start\".\n",
    "# The trinucleotide context is defined as the 3-mer sequence around the variant, where the variant is in the middle.\n",
    "# Add a new column \"tri_context\" to the DataFrame AML_variants.\n",
    "# Add also site-specific mutation rates to the DataFrame AML_variants.\n",
    "# The site-specific mutation rates are stored in the DataFrame site_mutation_rates.\n",
    "# The DataFrame site_mutation_rates has the following columns: \"Site\" and \"Mutation_rate\".\n",
    "AML_variants[\"tri_context\"] = \"\"\n",
    "AML_variants[\"site_specific_mutation_rate\"] = 0.0\n",
    "for index, row in AML_variants.iterrows():\n",
    "\tif row[\"exonic function\"] == \"nonsynonymous SNV\":\n",
    "\t\tchrom = \"chr\" + row[\"chr\"]\n",
    "\t\tstart = row[\"start\"] - 2\n",
    "\t\tend = row[\"start\"] + 1\n",
    "\t\tsequence = hg19_fasta[chrom][start:end]\n",
    "\t\tp0 = sequence[0]\n",
    "\t\tp1 = sequence[1]\n",
    "\t\tp2 = sequence[2]\n",
    "\t\ttri_context = f\"{p0}[{p1}>{row['alt allele']}]{p2}\"\n",
    "\t\ttri_rate = site_mutation_rates[site_mutation_rates[\"Site\"] == tri_context][\"Mutation_rate\"].values[0]\n",
    "\t\tprint(f\"ref = {row['ref allele']}, alt = {row['alt allele']}, queried = {sequence}, tri_context = {tri_context}, tri_rate = {tri_rate}\")\n",
    "\t\tAML_variants.at[index, \"tri_context\"] = tri_context\n",
    "\t\tAML_variants.at[index, \"site_specific_mutation_rate\"] = tri_rate\n",
    "\t\t"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {},
   "outputs": [],
   "source": [
    "# For all other variants, set the site-specific mutation rate to 2.7e-9.\n",
    "AML_variants[\"site_specific_mutation_rate\"] = AML_variants[\"site_specific_mutation_rate\"].replace(0.0, 2.7e-9)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample ID</th>\n",
       "      <th>gene</th>\n",
       "      <th>chr</th>\n",
       "      <th>start</th>\n",
       "      <th>ref allele</th>\n",
       "      <th>alt allele</th>\n",
       "      <th>amino acid change</th>\n",
       "      <th>function</th>\n",
       "      <th>exonic function</th>\n",
       "      <th>mutated %</th>\n",
       "      <th>scDNA-seq VAF</th>\n",
       "      <th>WT\\n(No. of cells)</th>\n",
       "      <th>Het\\n(No. of cells)</th>\n",
       "      <th>Homo\\n(No. of cells)</th>\n",
       "      <th>Missing\\n(No. of cells)</th>\n",
       "      <th>validated</th>\n",
       "      <th>validation method</th>\n",
       "      <th>bulk VAF</th>\n",
       "      <th>tri_context</th>\n",
       "      <th>site_specific_mutation_rate</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AML-01-001</td>\n",
       "      <td>FLT3</td>\n",
       "      <td>13</td>\n",
       "      <td>28608259</td>\n",
       "      <td>X</td>\n",
       "      <td>XAACTCTAAATTTTCTCTTGGAAACTCCCATTTGAGATCATATTCA...</td>\n",
       "      <td>p.Y599_D600delinsX</td>\n",
       "      <td>exonic</td>\n",
       "      <td>stopgain</td>\n",
       "      <td>0.935200</td>\n",
       "      <td>0.7949</td>\n",
       "      <td>239</td>\n",
       "      <td>8444</td>\n",
       "      <td>2190</td>\n",
       "      <td>498</td>\n",
       "      <td>yes</td>\n",
       "      <td>bulk-NGS</td>\n",
       "      <td>0.1186</td>\n",
       "      <td></td>\n",
       "      <td>2.700000e-09</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>AML-01-001</td>\n",
       "      <td>NPM1</td>\n",
       "      <td>5</td>\n",
       "      <td>170837544</td>\n",
       "      <td>T</td>\n",
       "      <td>TCTGC</td>\n",
       "      <td>p.L287fs</td>\n",
       "      <td>exonic</td>\n",
       "      <td>frameshift insertion</td>\n",
       "      <td>0.424900</td>\n",
       "      <td>0.4608</td>\n",
       "      <td>1285</td>\n",
       "      <td>4226</td>\n",
       "      <td>606</td>\n",
       "      <td>5254</td>\n",
       "      <td>yes</td>\n",
       "      <td>bulk-NGS</td>\n",
       "      <td>0.1828</td>\n",
       "      <td></td>\n",
       "      <td>2.700000e-09</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>AML-01-001</td>\n",
       "      <td>WT1</td>\n",
       "      <td>11</td>\n",
       "      <td>32417913</td>\n",
       "      <td>C</td>\n",
       "      <td>CGTACAAGA</td>\n",
       "      <td>p.R380fs</td>\n",
       "      <td>exonic</td>\n",
       "      <td>frameshift insertion</td>\n",
       "      <td>0.003400</td>\n",
       "      <td>0.0015</td>\n",
       "      <td>10779</td>\n",
       "      <td>39</td>\n",
       "      <td>0</td>\n",
       "      <td>553</td>\n",
       "      <td>yes</td>\n",
       "      <td>bulk-NGS in paired sample</td>\n",
       "      <td>0</td>\n",
       "      <td></td>\n",
       "      <td>2.700000e-09</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>AML-01-001</td>\n",
       "      <td>WT1</td>\n",
       "      <td>11</td>\n",
       "      <td>32417943</td>\n",
       "      <td>C</td>\n",
       "      <td>CG</td>\n",
       "      <td>p.R370fs</td>\n",
       "      <td>exonic</td>\n",
       "      <td>frameshift insertion</td>\n",
       "      <td>0.883600</td>\n",
       "      <td>0.4865</td>\n",
       "      <td>568</td>\n",
       "      <td>9799</td>\n",
       "      <td>248</td>\n",
       "      <td>756</td>\n",
       "      <td>yes</td>\n",
       "      <td>bulk-NGS</td>\n",
       "      <td>0.236</td>\n",
       "      <td></td>\n",
       "      <td>2.700000e-09</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>AML-02-001</td>\n",
       "      <td>DNMT3A</td>\n",
       "      <td>2</td>\n",
       "      <td>25457242</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>p.R882H</td>\n",
       "      <td>exonic</td>\n",
       "      <td>nonsynonymous SNV</td>\n",
       "      <td>0.760300</td>\n",
       "      <td>0.4639</td>\n",
       "      <td>468</td>\n",
       "      <td>5861</td>\n",
       "      <td>169</td>\n",
       "      <td>1433</td>\n",
       "      <td>yes</td>\n",
       "      <td>bulk-NGS</td>\n",
       "      <td>0.4239</td>\n",
       "      <td>G[C&gt;T]G</td>\n",
       "      <td>1.882290e-08</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>538</th>\n",
       "      <td>AML-122-001</td>\n",
       "      <td>SF3B1</td>\n",
       "      <td>2</td>\n",
       "      <td>198267359</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>p.K666N</td>\n",
       "      <td>exonic</td>\n",
       "      <td>nonsynonymous SNV</td>\n",
       "      <td>0.903884</td>\n",
       "      <td>0.5040</td>\n",
       "      <td>197</td>\n",
       "      <td>3497</td>\n",
       "      <td>133</td>\n",
       "      <td>189</td>\n",
       "      <td>yes</td>\n",
       "      <td>clinical sequencing</td>\n",
       "      <td>not available</td>\n",
       "      <td>T[C&gt;G]T</td>\n",
       "      <td>4.694750e-10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>539</th>\n",
       "      <td>AML-123-001</td>\n",
       "      <td>DNMT3A</td>\n",
       "      <td>2</td>\n",
       "      <td>25457243</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>p.R882C</td>\n",
       "      <td>exonic</td>\n",
       "      <td>nonsynonymous SNV</td>\n",
       "      <td>0.631245</td>\n",
       "      <td>0.3954</td>\n",
       "      <td>643</td>\n",
       "      <td>1495</td>\n",
       "      <td>97</td>\n",
       "      <td>287</td>\n",
       "      <td>yes</td>\n",
       "      <td>clinical sequencing</td>\n",
       "      <td>not available</td>\n",
       "      <td>C[G&gt;A]G</td>\n",
       "      <td>1.415120e-08</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>540</th>\n",
       "      <td>AML-123-001</td>\n",
       "      <td>RUNX1</td>\n",
       "      <td>21</td>\n",
       "      <td>36231774</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>p.R204X</td>\n",
       "      <td>exonic</td>\n",
       "      <td>stopgain</td>\n",
       "      <td>0.420698</td>\n",
       "      <td>0.2909</td>\n",
       "      <td>1061</td>\n",
       "      <td>1031</td>\n",
       "      <td>30</td>\n",
       "      <td>400</td>\n",
       "      <td>yes</td>\n",
       "      <td>clinical sequencing</td>\n",
       "      <td>not available</td>\n",
       "      <td></td>\n",
       "      <td>2.700000e-09</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>541</th>\n",
       "      <td>AML-123-001</td>\n",
       "      <td>TET2</td>\n",
       "      <td>4</td>\n",
       "      <td>106157248</td>\n",
       "      <td>C</td>\n",
       "      <td>CAA</td>\n",
       "      <td>p.H717fs</td>\n",
       "      <td>exonic</td>\n",
       "      <td>frameshift insertion</td>\n",
       "      <td>0.706186</td>\n",
       "      <td>0.3574</td>\n",
       "      <td>620</td>\n",
       "      <td>1751</td>\n",
       "      <td>30</td>\n",
       "      <td>121</td>\n",
       "      <td>yes</td>\n",
       "      <td>clinical sequencing</td>\n",
       "      <td>not available</td>\n",
       "      <td></td>\n",
       "      <td>2.700000e-09</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>542</th>\n",
       "      <td>AML-123-001</td>\n",
       "      <td>U2AF1</td>\n",
       "      <td>21</td>\n",
       "      <td>44524456</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>p.S34F</td>\n",
       "      <td>exonic</td>\n",
       "      <td>nonsynonymous SNV</td>\n",
       "      <td>0.649881</td>\n",
       "      <td>0.3488</td>\n",
       "      <td>734</td>\n",
       "      <td>1499</td>\n",
       "      <td>140</td>\n",
       "      <td>149</td>\n",
       "      <td>yes</td>\n",
       "      <td>clinical sequencing</td>\n",
       "      <td>not available</td>\n",
       "      <td>A[G&gt;A]A</td>\n",
       "      <td>1.396330e-09</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>530 rows × 20 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       sample ID    gene chr      start ref allele  \\\n",
       "0     AML-01-001    FLT3  13   28608259          X   \n",
       "1     AML-01-001    NPM1   5  170837544          T   \n",
       "2     AML-01-001     WT1  11   32417913          C   \n",
       "3     AML-01-001     WT1  11   32417943          C   \n",
       "4     AML-02-001  DNMT3A   2   25457242          C   \n",
       "..           ...     ...  ..        ...        ...   \n",
       "538  AML-122-001   SF3B1   2  198267359          C   \n",
       "539  AML-123-001  DNMT3A   2   25457243          G   \n",
       "540  AML-123-001   RUNX1  21   36231774          G   \n",
       "541  AML-123-001    TET2   4  106157248          C   \n",
       "542  AML-123-001   U2AF1  21   44524456          G   \n",
       "\n",
       "                                            alt allele   amino acid change  \\\n",
       "0    XAACTCTAAATTTTCTCTTGGAAACTCCCATTTGAGATCATATTCA...  p.Y599_D600delinsX   \n",
       "1                                                TCTGC            p.L287fs   \n",
       "2                                            CGTACAAGA            p.R380fs   \n",
       "3                                                   CG            p.R370fs   \n",
       "4                                                    T             p.R882H   \n",
       "..                                                 ...                 ...   \n",
       "538                                                  G             p.K666N   \n",
       "539                                                  A             p.R882C   \n",
       "540                                                  A             p.R204X   \n",
       "541                                                CAA            p.H717fs   \n",
       "542                                                  A              p.S34F   \n",
       "\n",
       "    function       exonic function  mutated %  scDNA-seq VAF  \\\n",
       "0     exonic              stopgain   0.935200         0.7949   \n",
       "1     exonic  frameshift insertion   0.424900         0.4608   \n",
       "2     exonic  frameshift insertion   0.003400         0.0015   \n",
       "3     exonic  frameshift insertion   0.883600         0.4865   \n",
       "4     exonic     nonsynonymous SNV   0.760300         0.4639   \n",
       "..       ...                   ...        ...            ...   \n",
       "538   exonic     nonsynonymous SNV   0.903884         0.5040   \n",
       "539   exonic     nonsynonymous SNV   0.631245         0.3954   \n",
       "540   exonic              stopgain   0.420698         0.2909   \n",
       "541   exonic  frameshift insertion   0.706186         0.3574   \n",
       "542   exonic     nonsynonymous SNV   0.649881         0.3488   \n",
       "\n",
       "     WT\\n(No. of cells)  Het\\n(No. of cells)  Homo\\n(No. of cells)  \\\n",
       "0                   239                 8444                  2190   \n",
       "1                  1285                 4226                   606   \n",
       "2                 10779                   39                     0   \n",
       "3                   568                 9799                   248   \n",
       "4                   468                 5861                   169   \n",
       "..                  ...                  ...                   ...   \n",
       "538                 197                 3497                   133   \n",
       "539                 643                 1495                    97   \n",
       "540                1061                 1031                    30   \n",
       "541                 620                 1751                    30   \n",
       "542                 734                 1499                   140   \n",
       "\n",
       "     Missing\\n(No. of cells) validated          validation method  \\\n",
       "0                        498       yes                   bulk-NGS   \n",
       "1                       5254       yes                   bulk-NGS   \n",
       "2                        553       yes  bulk-NGS in paired sample   \n",
       "3                        756       yes                   bulk-NGS   \n",
       "4                       1433       yes                   bulk-NGS   \n",
       "..                       ...       ...                        ...   \n",
       "538                      189       yes        clinical sequencing   \n",
       "539                      287       yes        clinical sequencing   \n",
       "540                      400       yes        clinical sequencing   \n",
       "541                      121       yes        clinical sequencing   \n",
       "542                      149       yes        clinical sequencing   \n",
       "\n",
       "          bulk VAF tri_context  site_specific_mutation_rate  \n",
       "0           0.1186                             2.700000e-09  \n",
       "1           0.1828                             2.700000e-09  \n",
       "2                0                             2.700000e-09  \n",
       "3            0.236                             2.700000e-09  \n",
       "4           0.4239     G[C>T]G                 1.882290e-08  \n",
       "..             ...         ...                          ...  \n",
       "538  not available     T[C>G]T                 4.694750e-10  \n",
       "539  not available     C[G>A]G                 1.415120e-08  \n",
       "540  not available                             2.700000e-09  \n",
       "541  not available                             2.700000e-09  \n",
       "542  not available     A[G>A]A                 1.396330e-09  \n",
       "\n",
       "[530 rows x 20 columns]"
      ]
     },
     "execution_count": 41,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "AML_variants"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Get all unique combinations of (gene, amino acid change, site_specific_mutation_rate)\n",
    "AML_variants_unique = AML_variants.drop_duplicates(subset=[\"gene\", \"amino acid change\", \"site_specific_mutation_rate\"])\n",
    "AML_variants_unique = AML_variants_unique[[\"gene\", \"amino acid change\", \"exonic function\", \"site_specific_mutation_rate\"]]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene</th>\n",
       "      <th>amino acid change</th>\n",
       "      <th>exonic function</th>\n",
       "      <th>site_specific_mutation_rate</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>NRAS</td>\n",
       "      <td>p.G12D</td>\n",
       "      <td>nonsynonymous SNV</td>\n",
       "      <td>3.175730e-09</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>NRAS</td>\n",
       "      <td>p.G13D</td>\n",
       "      <td>nonsynonymous SNV</td>\n",
       "      <td>3.175730e-09</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>NRAS</td>\n",
       "      <td>p.Q61H</td>\n",
       "      <td>nonsynonymous SNV</td>\n",
       "      <td>2.521550e-10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>NRAS</td>\n",
       "      <td>p.G13R</td>\n",
       "      <td>nonsynonymous SNV</td>\n",
       "      <td>4.643030e-10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>81</th>\n",
       "      <td>NRAS</td>\n",
       "      <td>p.G13V</td>\n",
       "      <td>nonsynonymous SNV</td>\n",
       "      <td>9.481170e-10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>88</th>\n",
       "      <td>NRAS</td>\n",
       "      <td>p.G13C</td>\n",
       "      <td>nonsynonymous SNV</td>\n",
       "      <td>1.538010e-09</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>89</th>\n",
       "      <td>NRAS</td>\n",
       "      <td>p.G12V</td>\n",
       "      <td>nonsynonymous SNV</td>\n",
       "      <td>9.481170e-10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>91</th>\n",
       "      <td>NRAS</td>\n",
       "      <td>p.G12C</td>\n",
       "      <td>nonsynonymous SNV</td>\n",
       "      <td>9.825080e-10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>92</th>\n",
       "      <td>NRAS</td>\n",
       "      <td>p.G12S</td>\n",
       "      <td>nonsynonymous SNV</td>\n",
       "      <td>4.379090e-09</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>162</th>\n",
       "      <td>NRAS</td>\n",
       "      <td>p.G12A</td>\n",
       "      <td>nonsynonymous SNV</td>\n",
       "      <td>4.832630e-10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>276</th>\n",
       "      <td>NRAS</td>\n",
       "      <td>p.Q61K</td>\n",
       "      <td>nonsynonymous SNV</td>\n",
       "      <td>1.332230e-09</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>311</th>\n",
       "      <td>NRAS</td>\n",
       "      <td>p.G60E</td>\n",
       "      <td>nonsynonymous SNV</td>\n",
       "      <td>1.560270e-09</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>316</th>\n",
       "      <td>NRAS</td>\n",
       "      <td>p.G12R</td>\n",
       "      <td>nonsynonymous SNV</td>\n",
       "      <td>4.370050e-10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>341</th>\n",
       "      <td>NRAS</td>\n",
       "      <td>p.Q61H</td>\n",
       "      <td>nonsynonymous SNV</td>\n",
       "      <td>3.835970e-10</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     gene amino acid change    exonic function  site_specific_mutation_rate\n",
       "14   NRAS            p.G12D  nonsynonymous SNV                 3.175730e-09\n",
       "25   NRAS            p.G13D  nonsynonymous SNV                 3.175730e-09\n",
       "27   NRAS            p.Q61H  nonsynonymous SNV                 2.521550e-10\n",
       "29   NRAS            p.G13R  nonsynonymous SNV                 4.643030e-10\n",
       "81   NRAS            p.G13V  nonsynonymous SNV                 9.481170e-10\n",
       "88   NRAS            p.G13C  nonsynonymous SNV                 1.538010e-09\n",
       "89   NRAS            p.G12V  nonsynonymous SNV                 9.481170e-10\n",
       "91   NRAS            p.G12C  nonsynonymous SNV                 9.825080e-10\n",
       "92   NRAS            p.G12S  nonsynonymous SNV                 4.379090e-09\n",
       "162  NRAS            p.G12A  nonsynonymous SNV                 4.832630e-10\n",
       "276  NRAS            p.Q61K  nonsynonymous SNV                 1.332230e-09\n",
       "311  NRAS            p.G60E  nonsynonymous SNV                 1.560270e-09\n",
       "316  NRAS            p.G12R  nonsynonymous SNV                 4.370050e-10\n",
       "341  NRAS            p.Q61H  nonsynonymous SNV                 3.835970e-10"
      ]
     },
     "execution_count": 54,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "AML_variants_unique.query(\"gene == 'NRAS'\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "metadata": {},
   "outputs": [],
   "source": [
    "# For each gene, sum up the site-specific mutation rates.\n",
    "AML_variants_sum = AML_variants.groupby(\"gene\").agg({\"site_specific_mutation_rate\": \"sum\"}).reset_index()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene</th>\n",
       "      <th>site_specific_mutation_rate</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ASXL1</td>\n",
       "      <td>4.590000e-08</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>BCOR</td>\n",
       "      <td>1.350000e-08</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>CBL</td>\n",
       "      <td>1.630361e-08</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>CSF3R</td>\n",
       "      <td>2.700000e-09</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>DNMT3A</td>\n",
       "      <td>6.364295e-07</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>ETV6</td>\n",
       "      <td>2.700000e-09</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>EZH2</td>\n",
       "      <td>5.053607e-08</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>FLT3</td>\n",
       "      <td>1.423273e-07</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>GATA2</td>\n",
       "      <td>6.238082e-09</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>IDH1</td>\n",
       "      <td>3.875233e-07</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>IDH2</td>\n",
       "      <td>4.279012e-07</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>JAK2</td>\n",
       "      <td>3.996690e-09</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>KIT</td>\n",
       "      <td>1.813268e-09</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>KRAS</td>\n",
       "      <td>7.434905e-08</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>MPL</td>\n",
       "      <td>1.415120e-08</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>MYC</td>\n",
       "      <td>1.577025e-08</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>NPM1</td>\n",
       "      <td>1.323000e-07</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>NRAS</td>\n",
       "      <td>1.557498e-07</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>PHF6</td>\n",
       "      <td>1.470440e-08</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>PPM1D</td>\n",
       "      <td>8.100000e-09</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>PTPN11</td>\n",
       "      <td>3.419075e-08</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>RUNX1</td>\n",
       "      <td>1.175633e-07</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>SETBP1</td>\n",
       "      <td>4.801760e-08</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>SF3B1</td>\n",
       "      <td>5.696345e-09</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>SMC3</td>\n",
       "      <td>4.832630e-10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>SRSF2</td>\n",
       "      <td>2.883850e-08</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>STAG2</td>\n",
       "      <td>1.080000e-08</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>TET2</td>\n",
       "      <td>6.021958e-08</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>TP53</td>\n",
       "      <td>5.109518e-08</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>U2AF1</td>\n",
       "      <td>9.089115e-09</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>WT1</td>\n",
       "      <td>1.049503e-07</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      gene  site_specific_mutation_rate\n",
       "0    ASXL1                 4.590000e-08\n",
       "1     BCOR                 1.350000e-08\n",
       "2      CBL                 1.630361e-08\n",
       "3    CSF3R                 2.700000e-09\n",
       "4   DNMT3A                 6.364295e-07\n",
       "5     ETV6                 2.700000e-09\n",
       "6     EZH2                 5.053607e-08\n",
       "7     FLT3                 1.423273e-07\n",
       "8    GATA2                 6.238082e-09\n",
       "9     IDH1                 3.875233e-07\n",
       "10    IDH2                 4.279012e-07\n",
       "11    JAK2                 3.996690e-09\n",
       "12     KIT                 1.813268e-09\n",
       "13    KRAS                 7.434905e-08\n",
       "14     MPL                 1.415120e-08\n",
       "15     MYC                 1.577025e-08\n",
       "16    NPM1                 1.323000e-07\n",
       "17    NRAS                 1.557498e-07\n",
       "18    PHF6                 1.470440e-08\n",
       "19   PPM1D                 8.100000e-09\n",
       "20  PTPN11                 3.419075e-08\n",
       "21   RUNX1                 1.175633e-07\n",
       "22  SETBP1                 4.801760e-08\n",
       "23   SF3B1                 5.696345e-09\n",
       "24    SMC3                 4.832630e-10\n",
       "25   SRSF2                 2.883850e-08\n",
       "26   STAG2                 1.080000e-08\n",
       "27    TET2                 6.021958e-08\n",
       "28    TP53                 5.109518e-08\n",
       "29   U2AF1                 9.089115e-09\n",
       "30     WT1                 1.049503e-07"
      ]
     },
     "execution_count": 57,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "AML_variants_sum"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 65,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "gene = FLT3, old mu = 1.4232733200000001e-07, new mu = 1.4232733200000001e-07\n",
      "gene = NPM1, old mu = 1.3230000000000001e-07, new mu = 1.3230000000000001e-07\n",
      "gene = WT1, old mu = 1.049503e-07, new mu = 1.049503e-07\n",
      "gene = DNMT3A, old mu = 6.36429535e-07, new mu = 6.36429535e-07\n",
      "gene = KRAS, old mu = 7.4349048e-08, new mu = 7.4349048e-08\n",
      "gene = NRAS, old mu = 1.55749763e-07, new mu = 1.55749763e-07\n",
      "gene = RUNX1, old mu = 1.1756328e-07, new mu = 1.1756328e-07\n",
      "gene = IDH1, old mu = 3.8752333000000003e-07, new mu = 3.8752333000000003e-07\n",
      "gene = IDH2, old mu = 4.2790116e-07, new mu = 4.2790116e-07\n",
      "gene = PTPN11, old mu = 3.4190748000000004e-08, new mu = 3.4190748000000004e-08\n",
      "gene = SRSF2, old mu = 2.88385e-08, new mu = 2.88385e-08\n",
      "gene = ASXL1, old mu = 4.59e-08, new mu = 4.59e-08\n",
      "gene = BCOR, old mu = 1.3500000000000002e-08, new mu = 1.3500000000000002e-08\n",
      "gene = STAG2, old mu = 1.08e-08, new mu = 1.08e-08\n",
      "gene = TP53, old mu = 5.1095175e-08, new mu = 5.1095175e-08\n",
      "gene = U2AF1, old mu = 9.089115e-09, new mu = 9.089115e-09\n",
      "gene = SF3B1, old mu = 5.696345e-09, new mu = 5.696345e-09\n",
      "gene = TET2, old mu = 6.021958100000001e-08, new mu = 6.021958100000001e-08\n",
      "gene = CSF3R, old mu = 2.7e-09, new mu = 2.7e-09\n",
      "gene = JAK2, old mu = 3.99669e-09, new mu = 3.99669e-09\n",
      "gene = GATA2, old mu = 6.238082000000001e-09, new mu = 6.238082000000001e-09\n",
      "gene = EZH2, old mu = 5.0536069e-08, new mu = 5.0536069e-08\n",
      "gene = PPM1D, old mu = 8.1e-09, new mu = 8.1e-09\n",
      "gene = SETBP1, old mu = 4.80176e-08, new mu = 4.80176e-08\n",
      "gene = KIT, old mu = 1.813268e-09, new mu = 1.813268e-09\n",
      "gene = CBL, old mu = 1.630361e-08, new mu = 1.630361e-08\n",
      "gene = PHF6, old mu = 1.4704399999999999e-08, new mu = 1.4704399999999999e-08\n",
      "gene = MYC, old mu = 1.5770254e-08, new mu = 1.5770254e-08\n",
      "gene = ETV6, old mu = 2.7e-09, new mu = 2.7e-09\n",
      "gene = MPL, old mu = 1.41512e-08, new mu = 1.41512e-08\n",
      "gene = SMC3, old mu = 4.83263e-10, new mu = 4.83263e-10\n"
     ]
    }
   ],
   "source": [
    "# Reset AML_cohort.mu_vec based on the calculated mutation rates\n",
    "for idx, gene in enumerate(AML_genes):\n",
    "\tnew_mu = AML_variants_sum[AML_variants_sum[\"gene\"] == gene][\"site_specific_mutation_rate\"].values[0]\n",
    "\tprint(f\"gene = {gene}, old mu = {AML_cohort.mu_vec[idx]}, new mu = {new_mu}\")\n",
    "\tAML_cohort.mu_vec[idx] = new_mu"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 66,
   "metadata": {},
   "outputs": [],
   "source": [
    "fitree.save_cohort_to_json(AML_cohort, \"../data/AML_morita/AML_cohort_Morita_2020.json\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "FiTree",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.9"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
